pkg/streamingpromql/types/ring_buffer.go

```diff
@@ -1,30 +1,80 @@
 // SPDX-License-Identifier: AGPL-3.0-only
-
 package types
 
 import (
 	"github.com/prometheus/prometheus/promql"
 )
 
-type RingBuffer struct {
-	pool       RingBufferPool
-	points     []promql.FPoint
+type RingBuffer[T any] struct {
+	pool       RingBufferPool[T]
+	points     []T
 	firstIndex int // Index into 'points' of first point in this buffer.
 	size       int // Number of points in this buffer.
 }
 
-type RingBufferPool interface {
+type RingBufferPool[T any] interface {
+	GetSlice(size int) ([]T, error)
+	PutSlice(s []T)
+	GetTimestamp(p T) int64
```

Comment:
`GetTimestamp` feels a bit out of place here - it doesn't have anything to do with pooling slices. 

Perhaps we could instead pass a function pointer to `RingBuffer`?

It'd also be good to understand the performance impact of the indirection - previously we were accessing `FPoint.T` directly, but now we need to go through a method, which has a small amount of overhead. Could you run the benchmarks for `rate()` over float samples before and after this change to see the impact? I'm not against the idea of having separate, specialised ring buffer types for `FPoint` and `HPoint` if need be.

Reply:
+1 unfortunately GO generics have terrible performance , especially in tight loops. Prometheus code is therefore full of non generic optimizations. This reminds me of https://github.com/prometheus/prometheus/blob/b6aba4ff144b893a9f19dea3eccd3cb6463752a7/storage/buffer.go#L271

Reply:
> GetTimestamp feels a bit out of place here - it doesn't have anything to do with pooling slices.

No, but the pool implementation is dependent on there being a timestamp so I think it make sense. Might be moot though if we don't use generics here.

So yes, it looks like there is a performance hit unfortunately.

<details>

```
goos: linux
goarch: amd64
pkg: github.com/grafana/mimir/pkg/streamingpromql/benchmarks
cpu: Intel(R) Xeon(R) CPU E5-2686 v4 @ 2.30GHz
                                                                                    │   main.txt   │   mqe-histogram-range-vectors.txt   │
                                                                                    │    sec/op    │    sec/op     vs base               │
Query/a_1,_instant_query/Mimir-8                                                       740.0µ ± 5%   684.3µ ±  5%   -7.53% (p=0.002 n=6)
Query/a_1,_instant_query/Prometheus-8                                                  753.7µ ± 3%   703.5µ ±  9%   -6.66% (p=0.002 n=6)
Query/a_1,_range_query_with_100_steps/Mimir-8                                          779.9µ ± 3%   709.3µ ±  9%   -9.05% (p=0.002 n=6)
Query/a_1,_range_query_with_100_steps/Prometheus-8                                     801.2µ ± 3%   749.7µ ±  3%   -6.43% (p=0.002 n=6)
Query/a_1,_range_query_with_1000_steps/Mimir-8                                         933.8µ ± 2%   909.6µ ±  2%   -2.59% (p=0.015 n=6)
Query/a_1,_range_query_with_1000_steps/Prometheus-8                                    938.8µ ± 3%   923.6µ ±  4%        ~ (p=0.310 n=6)
Query/a_100,_instant_query/Mimir-8                                                     1.914m ± 1%   1.889m ±  2%        ~ (p=0.485 n=6)
Query/a_100,_instant_query/Prometheus-8                                                1.947m ± 1%   1.919m ±  2%   -1.45% (p=0.026 n=6)
Query/a_100,_range_query_with_100_steps/Mimir-8                                        2.895m ± 2%   2.881m ±  2%        ~ (p=0.394 n=6)
Query/a_100,_range_query_with_100_steps/Prometheus-8                                   2.959m ± 2%   2.936m ±  2%        ~ (p=0.310 n=6)
Query/a_100,_range_query_with_1000_steps/Mimir-8                                       12.09m ± 1%   12.10m ±  1%        ~ (p=0.818 n=6)
Query/a_100,_range_query_with_1000_steps/Prometheus-8                                  12.17m ± 3%   12.38m ±  2%        ~ (p=0.310 n=6)
Query/a_2000,_instant_query/Mimir-8                                                    20.56m ± 3%   20.44m ±  2%        ~ (p=0.394 n=6)
Query/a_2000,_instant_query/Prometheus-8                                               21.05m ± 3%   20.83m ±  2%        ~ (p=0.310 n=6)
Query/a_2000,_range_query_with_100_steps/Mimir-8                                       40.53m ± 1%   40.57m ±  2%        ~ (p=1.000 n=6)
Query/a_2000,_range_query_with_100_steps/Prometheus-8                                  41.17m ± 2%   41.07m ±  1%        ~ (p=0.310 n=6)
Query/a_2000,_range_query_with_1000_steps/Mimir-8                                      223.9m ± 0%   222.7m ±  2%        ~ (p=0.485 n=6)
Query/a_2000,_range_query_with_1000_steps/Prometheus-8                                 227.7m ± 1%   228.2m ±  1%        ~ (p=0.589 n=6)
Query/nh_1,_instant_query/Mimir-8                                                      726.8µ ± 8%   778.2µ ±  2%   +7.07% (p=0.041 n=6)
Query/nh_1,_instant_query/Prometheus-8                                                 724.1µ ± 8%   794.6µ ±  3%   +9.74% (p=0.004 n=6)
Query/nh_1,_range_query_with_100_steps/Mimir-8                                         842.1µ ± 4%   872.2µ ±  3%   +3.57% (p=0.002 n=6)
Query/nh_1,_range_query_with_100_steps/Prometheus-8                                    881.8µ ± 3%   895.1µ ±  2%        ~ (p=0.485 n=6)
Query/nh_1,_range_query_with_1000_steps/Mimir-8                                        1.637m ± 2%   1.635m ±  2%        ~ (p=0.699 n=6)
Query/nh_1,_range_query_with_1000_steps/Prometheus-8                                   1.652m ± 2%   1.658m ±  1%        ~ (p=0.485 n=6)
Query/nh_100,_instant_query/Mimir-8                                                    3.923m ± 1%   3.917m ±  1%        ~ (p=0.818 n=6)
Query/nh_100,_instant_query/Prometheus-8                                               3.960m ± 1%   3.931m ±  1%        ~ (p=0.240 n=6)
Query/nh_100,_range_query_with_100_steps/Mimir-8                                       11.47m ± 1%   11.34m ±  3%        ~ (p=0.310 n=6)
Query/nh_100,_range_query_with_100_steps/Prometheus-8                                  11.59m ± 2%   11.48m ±  1%        ~ (p=0.180 n=6)
Query/nh_100,_range_query_with_1000_steps/Mimir-8                                      80.85m ± 1%   80.12m ±  2%        ~ (p=0.240 n=6)
Query/nh_100,_range_query_with_1000_steps/Prometheus-8                                 81.30m ± 1%   81.56m ±  2%        ~ (p=0.818 n=6)
Query/nh_2000,_instant_query/Mimir-8                                                   62.49m ± 1%   62.10m ±  2%        ~ (p=0.310 n=6)
Query/nh_2000,_instant_query/Prometheus-8                                              62.73m ± 1%   62.86m ±  1%        ~ (p=0.937 n=6)
Query/nh_2000,_range_query_with_100_steps/Mimir-8                                      213.8m ± 1%   212.2m ±  2%   -0.75% (p=0.026 n=6)
Query/nh_2000,_range_query_with_100_steps/Prometheus-8                                 215.8m ± 1%   214.9m ±  2%        ~ (p=0.589 n=6)
Query/nh_2000,_range_query_with_1000_steps/Mimir-8                                      1.949 ± 2%    1.934 ± 10%        ~ (p=0.485 n=6)
Query/nh_2000,_range_query_with_1000_steps/Prometheus-8                                 1.947 ± 4%    1.978 ±  2%        ~ (p=0.589 n=6)
Query/a_1[1m],_instant_query/Mimir-8                                                   690.6µ ± 6%   713.9µ ±  2%        ~ (p=0.065 n=6)
Query/a_1[1m],_instant_query/Prometheus-8                                              738.3µ ± 2%   752.6µ ±  5%        ~ (p=0.065 n=6)
Query/a_100[1m],_instant_query/Mimir-8                                                 1.178m ± 2%   1.181m ±  2%        ~ (p=0.818 n=6)
Query/a_100[1m],_instant_query/Prometheus-8                                            1.269m ± 2%   1.264m ±  2%        ~ (p=0.132 n=6)
Query/a_2000[1m],_instant_query/Mimir-8                                                8.026m ± 2%   8.016m ±  2%        ~ (p=0.485 n=6)
Query/a_2000[1m],_instant_query/Prometheus-8                                           9.425m ± 1%   9.462m ±  2%        ~ (p=0.818 n=6)
Query/rate(a_1[1m]),_instant_query/Mimir-8                                             717.2µ ± 3%   733.7µ ±  2%        ~ (p=0.240 n=6)
Query/rate(a_1[1m]),_instant_query/Prometheus-8                                        765.2µ ± 4%   775.0µ ±  2%        ~ (p=0.589 n=6)
Query/rate(a_1[1m]),_range_query_with_100_steps/Mimir-8                                751.1µ ± 2%   753.2µ ±  6%        ~ (p=0.394 n=6)
Query/rate(a_1[1m]),_range_query_with_100_steps/Prometheus-8                           796.2µ ± 4%   803.7µ ±  4%        ~ (p=0.937 n=6)
Query/rate(a_1[1m]),_range_query_with_1000_steps/Mimir-8                               915.8µ ± 2%   955.3µ ±  1%   +4.32% (p=0.002 n=6)
Query/rate(a_1[1m]),_range_query_with_1000_steps/Prometheus-8                          1.044m ± 2%   1.044m ±  2%        ~ (p=0.589 n=6)
Query/rate(a_100[1m]),_instant_query/Mimir-8                                           1.169m ± 2%   1.160m ±  2%        ~ (p=0.180 n=6)
Query/rate(a_100[1m]),_instant_query/Prometheus-8                                      1.364m ± 2%   1.376m ±  2%        ~ (p=0.818 n=6)
Query/rate(a_100[1m]),_range_query_with_100_steps/Mimir-8                              2.470m ± 1%   2.612m ±  1%   +5.78% (p=0.002 n=6)
Query/rate(a_100[1m]),_range_query_with_100_steps/Prometheus-8                         3.427m ± 1%   3.444m ±  2%        ~ (p=0.699 n=6)
Query/rate(a_100[1m]),_range_query_with_1000_steps/Mimir-8                             13.82m ± 1%   15.25m ±  3%  +10.39% (p=0.002 n=6)
Query/rate(a_100[1m]),_range_query_with_1000_steps/Prometheus-8                        23.03m ± 1%   23.02m ±  1%        ~ (p=0.818 n=6)
Query/rate(a_2000[1m]),_instant_query/Mimir-8                                          7.803m ± 1%   7.847m ±  1%        ~ (p=0.180 n=6)
Query/rate(a_2000[1m]),_instant_query/Prometheus-8                                     11.12m ± 2%   11.23m ±  1%        ~ (p=0.180 n=6)
Query/rate(a_2000[1m]),_range_query_with_100_steps/Mimir-8                             32.96m ± 2%   36.27m ±  1%  +10.05% (p=0.002 n=6)
Query/rate(a_2000[1m]),_range_query_with_100_steps/Prometheus-8                        54.85m ± 1%   54.63m ±  2%        ~ (p=0.937 n=6)
Query/rate(a_2000[1m]),_range_query_with_1000_steps/Mimir-8                            260.0m ± 1%   288.8m ±  1%  +11.06% (p=0.002 n=6)
Query/rate(a_2000[1m]),_range_query_with_1000_steps/Prometheus-8                       439.5m ± 1%   441.1m ±  1%        ~ (p=0.937 n=6)
Query/rate(a_1[1m]),_range_query_with_10000_steps/Mimir-8                              2.542m ± 2%   2.681m ±  1%   +5.45% (p=0.002 n=6)
Query/rate(a_1[1m]),_range_query_with_10000_steps/Prometheus-8                         3.244m ± 1%   3.268m ±  2%        ~ (p=1.000 n=6)
Query/rate(a_100[1m]),_range_query_with_10000_steps/Mimir-8                            138.8m ± 1%   153.2m ±  1%  +10.38% (p=0.002 n=6)
Query/rate(a_100[1m]),_range_query_with_10000_steps/Prometheus-8                       226.0m ± 1%   227.1m ±  1%        ~ (p=0.699 n=6)
Query/rate(a_2000[1m]),_range_query_with_10000_steps/Mimir-8                            2.747 ± 1%    3.054 ±  1%  +11.19% (p=0.002 n=6)
Query/rate(a_2000[1m]),_range_query_with_10000_steps/Prometheus-8                       4.736 ± 1%    4.689 ±  0%   -0.99% (p=0.026 n=6)
Query/rate(a_1[1d]),_instant_query/Mimir-8                                             2.096m ± 1%   2.090m ±  3%        ~ (p=0.485 n=6)
Query/rate(a_1[1d]),_instant_query/Prometheus-8                                        2.438m ± 3%   2.394m ±  1%        ~ (p=0.065 n=6)
Query/rate(a_1[1d]),_range_query_with_100_steps/Mimir-8                                2.777m ± 1%   2.802m ±  2%        ~ (p=0.485 n=6)
Query/rate(a_1[1d]),_range_query_with_100_steps/Prometheus-8                           3.233m ± 2%   3.228m ±  3%        ~ (p=0.589 n=6)
Query/rate(a_1[1d]),_range_query_with_1000_steps/Mimir-8                               9.053m ± 2%   9.104m ±  2%        ~ (p=0.093 n=6)
Query/rate(a_1[1d]),_range_query_with_1000_steps/Prometheus-8                          10.89m ± 2%   10.88m ±  1%        ~ (p=0.818 n=6)
Query/rate(a_100[1d]),_instant_query/Mimir-8                                           92.20m ± 1%   93.10m ±  1%        ~ (p=0.180 n=6)
Query/rate(a_100[1d]),_instant_query/Prometheus-8                                      109.1m ± 1%   109.2m ±  2%        ~ (p=0.699 n=6)
Query/rate(a_100[1d]),_range_query_with_100_steps/Mimir-8                              169.0m ± 1%   170.9m ±  1%   +1.08% (p=0.002 n=6)
Query/rate(a_100[1d]),_range_query_with_100_steps/Prometheus-8                         199.8m ± 1%   200.1m ±  1%        ~ (p=0.485 n=6)
Query/rate(a_100[1d]),_range_query_with_1000_steps/Mimir-8                             856.9m ± 1%   872.3m ±  0%   +1.80% (p=0.002 n=6)
Query/rate(a_100[1d]),_range_query_with_1000_steps/Prometheus-8                         1.021 ± 0%    1.026 ±  0%   +0.53% (p=0.009 n=6)
Query/rate(a_2000[1d]),_instant_query/Mimir-8                                           1.652 ± 1%    1.680 ±  1%   +1.68% (p=0.002 n=6)
Query/rate(a_2000[1d]),_instant_query/Prometheus-8                                      1.973 ± 1%    1.977 ±  1%        ~ (p=0.818 n=6)
Query/rate(a_2000[1d]),_range_query_with_100_steps/Mimir-8                              3.174 ± 1%    3.220 ±  1%   +1.47% (p=0.004 n=6)
Query/rate(a_2000[1d]),_range_query_with_100_steps/Prometheus-8                         3.804 ± 0%    3.792 ±  0%        ~ (p=0.065 n=6)
Query/rate(a_2000[1d]),_range_query_with_1000_steps/Mimir-8                             17.00 ± 0%    17.26 ±  1%   +1.53% (p=0.002 n=6)
Query/rate(a_2000[1d]),_range_query_with_1000_steps/Prometheus-8                        20.22 ± 1%    20.21 ±  0%        ~ (p=0.699 n=6)
Query/a_1_-_b_1,_instant_query/Mimir-8                                                 1.446m ± 3%   1.459m ±  3%        ~ (p=0.180 n=6)
Query/a_1_-_b_1,_instant_query/Prometheus-8                                            1.496m ± 4%   1.506m ±  2%        ~ (p=0.937 n=6)
Query/a_1_-_b_1,_range_query_with_100_steps/Mimir-8                                    1.508m ± 4%   1.500m ±  5%        ~ (p=0.818 n=6)
Query/a_1_-_b_1,_range_query_with_100_steps/Prometheus-8                               1.606m ± 3%   1.570m ±  2%   -2.26% (p=0.026 n=6)
Query/a_1_-_b_1,_range_query_with_1000_steps/Mimir-8                                   1.878m ± 1%   1.830m ±  4%        ~ (p=0.065 n=6)
Query/a_1_-_b_1,_range_query_with_1000_steps/Prometheus-8                              2.292m ± 3%   2.270m ±  1%        ~ (p=0.310 n=6)
Query/a_100_-_b_100,_instant_query/Mimir-8                                             3.903m ± 1%   3.888m ±  1%        ~ (p=0.180 n=6)
Query/a_100_-_b_100,_instant_query/Prometheus-8                                        3.867m ± 2%   3.922m ±  2%        ~ (p=0.180 n=6)
Query/a_100_-_b_100,_range_query_with_100_steps/Mimir-8                                5.877m ± 2%   5.877m ±  3%        ~ (p=0.937 n=6)
Query/a_100_-_b_100,_range_query_with_100_steps/Prometheus-8                           9.324m ± 2%   9.433m ±  3%   +1.17% (p=0.041 n=6)
Query/a_100_-_b_100,_range_query_with_1000_steps/Mimir-8                               25.21m ± 2%   24.84m ±  1%        ~ (p=0.065 n=6)
Query/a_100_-_b_100,_range_query_with_1000_steps/Prometheus-8                          66.01m ± 0%   65.83m ±  1%        ~ (p=0.699 n=6)
Query/a_2000_-_b_2000,_instant_query/Mimir-8                                           45.03m ± 1%   45.56m ±  1%   +1.18% (p=0.026 n=6)
Query/a_2000_-_b_2000,_instant_query/Prometheus-8                                      45.76m ± 2%   45.24m ±  2%        ~ (p=0.132 n=6)
Query/a_2000_-_b_2000,_range_query_with_100_steps/Mimir-8                              86.36m ± 2%   85.69m ±  2%        ~ (p=1.000 n=6)
Query/a_2000_-_b_2000,_range_query_with_100_steps/Prometheus-8                         177.7m ± 2%   178.7m ±  1%        ~ (p=0.310 n=6)
Query/a_2000_-_b_2000,_range_query_with_1000_steps/Mimir-8                             462.6m ± 1%   458.8m ±  1%   -0.82% (p=0.009 n=6)
Query/a_2000_-_b_2000,_range_query_with_1000_steps/Prometheus-8                         1.458 ± 1%    1.464 ±  1%        ~ (p=0.240 n=6)
Query/a_1_-_b_1,_range_query_with_10000_steps/Mimir-8                                  4.658m ± 3%   4.521m ±  3%   -2.94% (p=0.026 n=6)
Query/a_1_-_b_1,_range_query_with_10000_steps/Prometheus-8                             8.379m ± 1%   8.355m ±  3%        ~ (p=0.818 n=6)
Query/a_100_-_b_100,_range_query_with_10000_steps/Mimir-8                              228.3m ± 0%   228.1m ±  1%        ~ (p=0.818 n=6)
Query/a_100_-_b_100,_range_query_with_10000_steps/Prometheus-8                         635.5m ± 1%   644.0m ±  0%   +1.34% (p=0.002 n=6)
Query/a_2000_-_b_2000,_range_query_with_10000_steps/Mimir-8                             4.510 ± 1%    4.491 ±  1%   -0.41% (p=0.041 n=6)
Query/a_2000_-_b_2000,_range_query_with_10000_steps/Prometheus-8                        15.10 ± 0%    15.21 ±  0%   +0.70% (p=0.002 n=6)
Query/a_100{l=~"[13579]."}_-_b_100,_instant_query/Mimir-8                              3.783m ± 2%   3.631m ±  2%   -4.04% (p=0.002 n=6)
Query/a_100{l=~"[13579]."}_-_b_100,_instant_query/Prometheus-8                         3.764m ± 2%   3.627m ±  2%   -3.63% (p=0.002 n=6)
Query/a_100{l=~"[13579]."}_-_b_100,_range_query_with_100_steps/Mimir-8                 5.515m ± 2%   5.256m ±  3%   -4.70% (p=0.009 n=6)
Query/a_100{l=~"[13579]."}_-_b_100,_range_query_with_100_steps/Prometheus-8            7.553m ± 0%   7.336m ±  1%   -2.87% (p=0.002 n=6)
Query/a_100{l=~"[13579]."}_-_b_100,_range_query_with_1000_steps/Mimir-8                21.61m ± 2%   21.55m ±  1%        ~ (p=0.240 n=6)
Query/a_100{l=~"[13579]."}_-_b_100,_range_query_with_1000_steps/Prometheus-8           45.85m ± 2%   45.86m ±  2%        ~ (p=0.818 n=6)
Query/a_2000{l=~"1..."}_-_b_2000,_instant_query/Mimir-8                                29.93m ± 3%   29.46m ±  2%   -1.59% (p=0.026 n=6)
Query/a_2000{l=~"1..."}_-_b_2000,_instant_query/Prometheus-8                           34.62m ± 1%   34.62m ±  3%        ~ (p=1.000 n=6)
Query/a_2000{l=~"1..."}_-_b_2000,_range_query_with_100_steps/Mimir-8                   53.64m ± 2%   52.89m ±  2%        ~ (p=0.180 n=6)
Query/a_2000{l=~"1..."}_-_b_2000,_range_query_with_100_steps/Prometheus-8              116.5m ± 1%   117.4m ±  2%   +0.74% (p=0.015 n=6)
Query/a_2000{l=~"1..."}_-_b_2000,_range_query_with_1000_steps/Mimir-8                  282.4m ± 1%   279.5m ±  2%   -1.02% (p=0.041 n=6)
Query/a_2000{l=~"1..."}_-_b_2000,_range_query_with_1000_steps/Prometheus-8             875.3m ± 1%   877.8m ±  0%        ~ (p=0.093 n=6)
Query/a_2000{l="1234"}_-_b_2000,_instant_query/Mimir-8                                 12.99m ± 1%   12.74m ±  1%   -1.88% (p=0.002 n=6)
Query/a_2000{l="1234"}_-_b_2000,_instant_query/Prometheus-8                            22.75m ± 1%   22.61m ±  2%        ~ (p=0.310 n=6)
Query/a_2000{l="1234"}_-_b_2000,_range_query_with_100_steps/Mimir-8                    19.42m ± 2%   19.19m ±  2%   -1.19% (p=0.041 n=6)
Query/a_2000{l="1234"}_-_b_2000,_range_query_with_100_steps/Prometheus-8               60.85m ± 2%   61.35m ±  2%        ~ (p=0.240 n=6)
Query/a_2000{l="1234"}_-_b_2000,_range_query_with_1000_steps/Mimir-8                   86.04m ± 3%   85.68m ±  2%        ~ (p=0.589 n=6)
Query/a_2000{l="1234"}_-_b_2000,_range_query_with_1000_steps/Prometheus-8              420.9m ± 1%   420.9m ±  1%        ~ (p=0.818 n=6)
Query/a_2000_-_b_2000{l="1234"},_instant_query/Mimir-8                                 12.72m ± 2%   12.55m ±  1%   -1.34% (p=0.026 n=6)
Query/a_2000_-_b_2000{l="1234"},_instant_query/Prometheus-8                            22.13m ± 2%   21.58m ±  1%   -2.52% (p=0.004 n=6)
Query/a_2000_-_b_2000{l="1234"},_range_query_with_100_steps/Mimir-8                    19.08m ± 2%   18.67m ±  2%   -2.13% (p=0.026 n=6)
Query/a_2000_-_b_2000{l="1234"},_range_query_with_100_steps/Prometheus-8               48.30m ± 2%   48.51m ±  3%        ~ (p=1.000 n=6)
Query/a_2000_-_b_2000{l="1234"},_range_query_with_1000_steps/Mimir-8                   83.75m ± 3%   84.34m ±  1%        ~ (p=0.699 n=6)
Query/a_2000_-_b_2000{l="1234"},_range_query_with_1000_steps/Prometheus-8              305.5m ± 1%   306.7m ±  2%        ~ (p=0.180 n=6)
Query/sum(a_1),_instant_query/Mimir-8                                                  766.2µ ± 4%   655.6µ ±  3%  -14.43% (p=0.002 n=6)
Query/sum(a_1),_instant_query/Prometheus-8                                             783.9µ ± 2%   670.3µ ±  4%  -14.50% (p=0.002 n=6)
Query/sum(a_1),_range_query_with_100_steps/Mimir-8                                     780.5µ ± 5%   676.5µ ±  2%  -13.33% (p=0.002 n=6)
Query/sum(a_1),_range_query_with_100_steps/Prometheus-8                                826.7µ ± 3%   726.9µ ±  4%  -12.07% (p=0.002 n=6)
Query/sum(a_1),_range_query_with_1000_steps/Mimir-8                                    949.8µ ± 2%   868.2µ ±  3%   -8.59% (p=0.002 n=6)
Query/sum(a_1),_range_query_with_1000_steps/Prometheus-8                              1015.0µ ± 2%   929.9µ ±  3%   -8.38% (p=0.002 n=6)
Query/sum(a_100),_instant_query/Mimir-8                                                1.913m ± 1%   1.857m ±  2%   -2.91% (p=0.002 n=6)
Query/sum(a_100),_instant_query/Prometheus-8                                           1.952m ± 1%   1.890m ±  2%   -3.17% (p=0.002 n=6)
Query/sum(a_100),_range_query_with_100_steps/Mimir-8                                   2.935m ± 1%   2.858m ±  1%   -2.63% (p=0.002 n=6)
Query/sum(a_100),_range_query_with_100_steps/Prometheus-8                              3.002m ± 1%   2.946m ±  3%   -1.87% (p=0.002 n=6)
Query/sum(a_100),_range_query_with_1000_steps/Mimir-8                                  12.31m ± 3%   12.21m ±  1%        ~ (p=0.394 n=6)
Query/sum(a_100),_range_query_with_1000_steps/Prometheus-8                             12.81m ± 2%   12.74m ±  2%        ~ (p=0.485 n=6)
Query/sum(a_2000),_instant_query/Mimir-8                                               20.94m ± 1%   20.60m ±  2%   -1.61% (p=0.026 n=6)
Query/sum(a_2000),_instant_query/Prometheus-8                                          20.98m ± 2%   20.88m ±  2%        ~ (p=0.485 n=6)
Query/sum(a_2000),_range_query_with_100_steps/Mimir-8                                  40.66m ± 2%   40.17m ±  2%        ~ (p=0.240 n=6)
Query/sum(a_2000),_range_query_with_100_steps/Prometheus-8                             42.89m ± 3%   42.79m ±  2%        ~ (p=0.937 n=6)
Query/sum(a_2000),_range_query_with_1000_steps/Mimir-8                                 228.2m ± 1%   227.6m ±  1%        ~ (p=0.180 n=6)
Query/sum(a_2000),_range_query_with_1000_steps/Prometheus-8                            257.8m ± 1%   258.8m ±  1%        ~ (p=0.485 n=6)
Query/sum_by_(l)(h_1),_instant_query/Mimir-8                                           843.1µ ± 3%   761.3µ ±  2%   -9.70% (p=0.002 n=6)
Query/sum_by_(l)(h_1),_instant_query/Prometheus-8                                      863.0µ ± 3%   794.0µ ±  4%   -8.00% (p=0.002 n=6)
Query/sum_by_(l)(h_1),_range_query_with_100_steps/Mimir-8                              921.3µ ± 3%   854.2µ ±  6%   -7.29% (p=0.002 n=6)
Query/sum_by_(l)(h_1),_range_query_with_100_steps/Prometheus-8                         964.1µ ± 2%   943.5µ ±  1%   -2.14% (p=0.015 n=6)
Query/sum_by_(l)(h_1),_range_query_with_1000_steps/Mimir-8                             1.564m ± 3%   1.539m ±  2%        ~ (p=0.065 n=6)
Query/sum_by_(l)(h_1),_range_query_with_1000_steps/Prometheus-8                        1.627m ± 2%   1.603m ±  2%        ~ (p=0.093 n=6)
Query/sum_by_(l)(h_100),_instant_query/Mimir-8                                         7.549m ± 1%   7.522m ±  3%        ~ (p=0.310 n=6)
Query/sum_by_(l)(h_100),_instant_query/Prometheus-8                                    7.660m ± 1%   7.689m ±  1%        ~ (p=0.485 n=6)
Query/sum_by_(l)(h_100),_range_query_with_100_steps/Mimir-8                            13.71m ± 3%   13.49m ±  2%        ~ (p=0.132 n=6)
Query/sum_by_(l)(h_100),_range_query_with_100_steps/Prometheus-8                       14.22m ± 2%   14.11m ±  2%        ~ (p=0.485 n=6)
Query/sum_by_(l)(h_100),_range_query_with_1000_steps/Mimir-8                           74.35m ± 2%   73.44m ±  2%        ~ (p=0.180 n=6)
Query/sum_by_(l)(h_100),_range_query_with_1000_steps/Prometheus-8                      83.94m ± 1%   84.05m ±  2%        ~ (p=0.937 n=6)
Query/sum_by_(l)(h_2000),_instant_query/Mimir-8                                        134.3m ± 1%   133.2m ±  1%        ~ (p=0.310 n=6)
Query/sum_by_(l)(h_2000),_instant_query/Prometheus-8                                   136.1m ± 2%   136.2m ±  1%        ~ (p=0.699 n=6)
Query/sum_by_(l)(h_2000),_range_query_with_100_steps/Mimir-8                           257.8m ± 2%   256.0m ±  2%        ~ (p=0.180 n=6)
Query/sum_by_(l)(h_2000),_range_query_with_100_steps/Prometheus-8                      280.1m ± 2%   278.5m ±  1%        ~ (p=0.310 n=6)
Query/sum_by_(l)(h_2000),_range_query_with_1000_steps/Mimir-8                           1.448 ± 1%    1.434 ±  0%   -0.99% (p=0.026 n=6)
Query/sum_by_(l)(h_2000),_range_query_with_1000_steps/Prometheus-8                      1.917 ± 1%    1.923 ±  1%        ~ (p=0.132 n=6)
Query/sum_by_(le)(h_1),_instant_query/Mimir-8                                          850.6µ ± 2%   854.6µ ±  4%        ~ (p=0.394 n=6)
Query/sum_by_(le)(h_1),_instant_query/Prometheus-8                                     876.2µ ± 3%   873.9µ ±  2%        ~ (p=0.937 n=6)
Query/sum_by_(le)(h_1),_range_query_with_100_steps/Mimir-8                             943.1µ ± 3%   918.9µ ±  3%        ~ (p=0.093 n=6)
Query/sum_by_(le)(h_1),_range_query_with_100_steps/Prometheus-8                        952.4µ ± 3%   962.4µ ±  2%        ~ (p=0.589 n=6)
Query/sum_by_(le)(h_1),_range_query_with_1000_steps/Mimir-8                            1.569m ± 1%   1.577m ±  1%        ~ (p=0.394 n=6)
Query/sum_by_(le)(h_1),_range_query_with_1000_steps/Prometheus-8                       1.665m ± 1%   1.685m ±  1%   +1.24% (p=0.041 n=6)
Query/sum_by_(le)(h_100),_instant_query/Mimir-8                                        7.498m ± 2%   7.517m ±  2%        ~ (p=0.818 n=6)
Query/sum_by_(le)(h_100),_instant_query/Prometheus-8                                   7.633m ± 3%   7.586m ±  1%        ~ (p=0.818 n=6)
Query/sum_by_(le)(h_100),_range_query_with_100_steps/Mimir-8                           13.42m ± 2%   13.39m ±  2%        ~ (p=0.699 n=6)
Query/sum_by_(le)(h_100),_range_query_with_100_steps/Prometheus-8                      13.97m ± 2%   13.86m ±  2%        ~ (p=0.394 n=6)
Query/sum_by_(le)(h_100),_range_query_with_1000_steps/Mimir-8                          73.89m ± 1%   73.37m ±  1%   -0.70% (p=0.026 n=6)
Query/sum_by_(le)(h_100),_range_query_with_1000_steps/Prometheus-8                     80.12m ± 1%   81.16m ±  1%        ~ (p=0.065 n=6)
Query/sum_by_(le)(h_2000),_instant_query/Mimir-8                                       131.8m ± 1%   131.4m ±  1%        ~ (p=0.699 n=6)
Query/sum_by_(le)(h_2000),_instant_query/Prometheus-8                                  134.8m ± 1%   134.8m ±  1%        ~ (p=0.589 n=6)
Query/sum_by_(le)(h_2000),_range_query_with_100_steps/Mimir-8                          257.4m ± 1%   253.1m ±  2%   -1.67% (p=0.015 n=6)
Query/sum_by_(le)(h_2000),_range_query_with_100_steps/Prometheus-8                     271.1m ± 1%   270.0m ±  1%        ~ (p=0.485 n=6)
Query/sum_by_(le)(h_2000),_range_query_with_1000_steps/Mimir-8                          1.425 ± 1%    1.406 ±  1%   -1.35% (p=0.002 n=6)
Query/sum_by_(le)(h_2000),_range_query_with_1000_steps/Prometheus-8                     1.785 ± 1%    1.798 ±  1%        ~ (p=0.132 n=6)
Query/rate(a_1[1m])_+_rate(b_1[1m]),_instant_query/Mimir-8                             1.437m ± 4%   1.422m ±  3%        ~ (p=0.485 n=6)
Query/rate(a_1[1m])_+_rate(b_1[1m]),_instant_query/Prometheus-8                        1.496m ± 3%   1.460m ±  4%        ~ (p=0.065 n=6)
Query/rate(a_1[1m])_+_rate(b_1[1m]),_range_query_with_100_steps/Mimir-8                1.459m ± 4%   1.421m ±  3%        ~ (p=0.093 n=6)
Query/rate(a_1[1m])_+_rate(b_1[1m]),_range_query_with_100_steps/Prometheus-8           1.591m ± 3%   1.557m ±  3%        ~ (p=0.180 n=6)
Query/rate(a_1[1m])_+_rate(b_1[1m]),_range_query_with_1000_steps/Mimir-8               1.846m ± 3%   1.867m ±  4%        ~ (p=0.180 n=6)
Query/rate(a_1[1m])_+_rate(b_1[1m]),_range_query_with_1000_steps/Prometheus-8          2.384m ± 3%   2.429m ±  3%        ~ (p=0.310 n=6)
Query/rate(a_100[1m])_+_rate(b_100[1m]),_instant_query/Mimir-8                         2.452m ± 1%   2.466m ±  1%        ~ (p=0.180 n=6)
Query/rate(a_100[1m])_+_rate(b_100[1m]),_instant_query/Prometheus-8                    2.759m ± 2%   2.782m ±  2%        ~ (p=0.394 n=6)
Query/rate(a_100[1m])_+_rate(b_100[1m]),_range_query_with_100_steps/Mimir-8            5.013m ± 1%   5.350m ±  2%   +6.70% (p=0.002 n=6)
Query/rate(a_100[1m])_+_rate(b_100[1m]),_range_query_with_100_steps/Prometheus-8       10.27m ± 3%   10.32m ±  2%        ~ (p=0.589 n=6)
Query/rate(a_100[1m])_+_rate(b_100[1m]),_range_query_with_1000_steps/Mimir-8           28.41m ± 1%   31.42m ±  1%  +10.59% (p=0.002 n=6)
Query/rate(a_100[1m])_+_rate(b_100[1m]),_range_query_with_1000_steps/Prometheus-8      86.24m ± 1%   86.27m ±  1%        ~ (p=0.818 n=6)
Query/rate(a_2000[1m])_+_rate(b_2000[1m]),_instant_query/Mimir-8                       18.19m ± 1%   18.58m ±  4%        ~ (p=0.093 n=6)
Query/rate(a_2000[1m])_+_rate(b_2000[1m]),_instant_query/Prometheus-8                  25.02m ± 1%   24.95m ±  1%        ~ (p=0.240 n=6)
Query/rate(a_2000[1m])_+_rate(b_2000[1m]),_range_query_with_100_steps/Mimir-8          71.03m ± 1%   76.82m ±  3%   +8.14% (p=0.002 n=6)
Query/rate(a_2000[1m])_+_rate(b_2000[1m]),_range_query_with_100_steps/Prometheus-8     202.8m ± 1%   202.9m ±  1%        ~ (p=0.240 n=6)
Query/rate(a_2000[1m])_+_rate(b_2000[1m]),_range_query_with_1000_steps/Mimir-8         536.1m ± 1%   592.0m ±  1%  +10.42% (p=0.002 n=6)
Query/rate(a_2000[1m])_+_rate(b_2000[1m]),_range_query_with_1000_steps/Prometheus-8     1.861 ± 1%    1.858 ±  1%        ~ (p=0.310 n=6)
Query/sum(a_1_+_b_1),_instant_query/Mimir-8                                            1.410m ± 5%   1.478m ±  5%   +4.76% (p=0.026 n=6)
Query/sum(a_1_+_b_1),_instant_query/Prometheus-8                                       1.435m ± 5%   1.517m ±  2%   +5.73% (p=0.009 n=6)
Query/sum(a_1_+_b_1),_range_query_with_100_steps/Mimir-8                               1.484m ± 4%   1.505m ±  3%        ~ (p=0.132 n=6)
Query/sum(a_1_+_b_1),_range_query_with_100_steps/Prometheus-8                          1.592m ± 5%   1.539m ±  2%   -3.33% (p=0.009 n=6)
Query/sum(a_1_+_b_1),_range_query_with_1000_steps/Mimir-8                              1.879m ± 1%   1.829m ±  3%        ~ (p=0.093 n=6)
Query/sum(a_1_+_b_1),_range_query_with_1000_steps/Prometheus-8                         2.308m ± 2%   2.304m ±  4%        ~ (p=0.818 n=6)
Query/sum(a_100_+_b_100),_instant_query/Mimir-8                                        3.909m ± 3%   3.925m ±  1%        ~ (p=0.394 n=6)
Query/sum(a_100_+_b_100),_instant_query/Prometheus-8                                   3.859m ± 2%   3.890m ±  2%        ~ (p=0.132 n=6)
Query/sum(a_100_+_b_100),_range_query_with_100_steps/Mimir-8                           5.963m ± 2%   5.881m ±  1%        ~ (p=0.065 n=6)
Query/sum(a_100_+_b_100),_range_query_with_100_steps/Prometheus-8                      9.466m ± 2%   9.429m ±  2%        ~ (p=0.485 n=6)
Query/sum(a_100_+_b_100),_range_query_with_1000_steps/Mimir-8                          25.52m ± 1%   25.02m ±  1%   -1.97% (p=0.002 n=6)
Query/sum(a_100_+_b_100),_range_query_with_1000_steps/Prometheus-8                     67.32m ± 2%   67.76m ±  0%        ~ (p=0.180 n=6)
Query/sum(a_2000_+_b_2000),_instant_query/Mimir-8                                      45.87m ± 2%   45.10m ±  2%        ~ (p=0.065 n=6)
Query/sum(a_2000_+_b_2000),_instant_query/Prometheus-8                                 46.23m ± 1%   45.50m ±  3%        ~ (p=0.240 n=6)
Query/sum(a_2000_+_b_2000),_range_query_with_100_steps/Mimir-8                         87.34m ± 1%   85.36m ±  2%   -2.26% (p=0.015 n=6)
Query/sum(a_2000_+_b_2000),_range_query_with_100_steps/Prometheus-8                    181.9m ± 1%   180.9m ±  2%        ~ (p=0.818 n=6)
Query/sum(a_2000_+_b_2000),_range_query_with_1000_steps/Mimir-8                        470.3m ± 0%   462.1m ±  1%   -1.75% (p=0.002 n=6)
Query/sum(a_2000_+_b_2000),_range_query_with_1000_steps/Prometheus-8                    1.494 ± 0%    1.484 ±  2%        ~ (p=0.065 n=6)
Query/sum_by_(le)(rate(h_1[1m])),_instant_query/Mimir-8                                782.4µ ± 3%   790.5µ ±  3%        ~ (p=0.818 n=6)
Query/sum_by_(le)(rate(h_1[1m])),_instant_query/Prometheus-8                           812.5µ ± 2%   822.1µ ±  4%        ~ (p=0.485 n=6)
Query/sum_by_(le)(rate(h_1[1m])),_range_query_with_100_steps/Mimir-8                   905.7µ ± 3%   912.9µ ±  2%        ~ (p=0.310 n=6)
Query/sum_by_(le)(rate(h_1[1m])),_range_query_with_100_steps/Prometheus-8             1010.8µ ± 3%   998.3µ ±  2%        ~ (p=0.240 n=6)
Query/sum_by_(le)(rate(h_1[1m])),_range_query_with_1000_steps/Mimir-8                  1.672m ± 2%   1.731m ±  2%   +3.57% (p=0.002 n=6)
Query/sum_by_(le)(rate(h_1[1m])),_range_query_with_1000_steps/Prometheus-8             2.249m ± 0%   2.254m ±  1%        ~ (p=0.485 n=6)
Query/sum_by_(le)(rate(h_100[1m])),_instant_query/Mimir-8                              3.341m ± 3%   3.314m ±  1%        ~ (p=0.310 n=6)
Query/sum_by_(le)(rate(h_100[1m])),_instant_query/Prometheus-8                         4.298m ± 3%   4.301m ±  1%        ~ (p=0.937 n=6)
Query/sum_by_(le)(rate(h_100[1m])),_range_query_with_100_steps/Mimir-8                 10.94m ± 3%   11.79m ±  3%   +7.84% (p=0.002 n=6)
Query/sum_by_(le)(rate(h_100[1m])),_range_query_with_100_steps/Prometheus-8            17.97m ± 3%   17.95m ±  1%        ~ (p=0.937 n=6)
Query/sum_by_(le)(rate(h_100[1m])),_range_query_with_1000_steps/Mimir-8                83.91m ± 1%   92.91m ±  2%  +10.74% (p=0.002 n=6)
Query/sum_by_(le)(rate(h_100[1m])),_range_query_with_1000_steps/Prometheus-8           146.0m ± 1%   146.0m ±  1%        ~ (p=0.589 n=6)
Query/sum_by_(le)(rate(h_2000[1m])),_instant_query/Mimir-8                             48.26m ± 3%   48.43m ±  1%        ~ (p=0.394 n=6)
Query/sum_by_(le)(rate(h_2000[1m])),_instant_query/Prometheus-8                        69.36m ± 2%   69.64m ±  3%        ~ (p=0.818 n=6)
Query/sum_by_(le)(rate(h_2000[1m])),_range_query_with_100_steps/Mimir-8                206.0m ± 1%   223.7m ±  0%   +8.63% (p=0.002 n=6)
Query/sum_by_(le)(rate(h_2000[1m])),_range_query_with_100_steps/Prometheus-8           351.1m ± 1%   351.3m ±  1%        ~ (p=1.000 n=6)
Query/sum_by_(le)(rate(h_2000[1m])),_range_query_with_1000_steps/Mimir-8                1.606 ± 1%    1.799 ±  1%  +12.01% (p=0.002 n=6)
Query/sum_by_(le)(rate(h_2000[1m])),_range_query_with_1000_steps/Prometheus-8           3.052 ± 1%    3.058 ±  1%        ~ (p=0.394 n=6)
geomean                                                                                18.12m        18.10m         -0.13%

                                                                                    │    main.txt    │   mqe-histogram-range-vectors.txt    │
                                                                                    │      B/op      │      B/op       vs base              │
Query/a_1,_instant_query/Mimir-8                                                       19.31Ki ±  0%    19.31Ki ±  0%       ~ (p=0.734 n=6)
Query/a_1,_instant_query/Prometheus-8                                                  20.84Ki ±  0%    20.81Ki ±  0%       ~ (p=0.240 n=6)
Query/a_1,_range_query_with_100_steps/Mimir-8                                          19.88Ki ±  0%    19.89Ki ±  0%       ~ (p=0.515 n=6)
Query/a_1,_range_query_with_100_steps/Prometheus-8                                     21.90Ki ±  0%    21.86Ki ±  2%       ~ (p=0.121 n=6)
Query/a_1,_range_query_with_1000_steps/Mimir-8                                         24.57Ki ±  1%    24.49Ki ±  1%       ~ (p=0.394 n=6)
Query/a_1,_range_query_with_1000_steps/Prometheus-8                                    26.62Ki ±  1%    26.60Ki ±  0%       ~ (p=0.240 n=6)
Query/a_100,_instant_query/Mimir-8                                                     156.8Ki ±  0%    156.9Ki ±  0%       ~ (p=0.937 n=6)
Query/a_100,_instant_query/Prometheus-8                                                173.3Ki ±  0%    173.5Ki ±  0%  +0.09% (p=0.041 n=6)
Query/a_100,_range_query_with_100_steps/Mimir-8                                        213.4Ki ±  0%    213.8Ki ±  0%       ~ (p=0.310 n=6)
Query/a_100,_range_query_with_100_steps/Prometheus-8                                   225.3Ki ±  0%    226.2Ki ±  0%       ~ (p=0.065 n=6)
Query/a_100,_range_query_with_1000_steps/Mimir-8                                       662.6Ki ±  1%    662.5Ki ±  1%       ~ (p=0.818 n=6)
Query/a_100,_range_query_with_1000_steps/Prometheus-8                                  673.7Ki ±  0%    673.4Ki ±  0%       ~ (p=0.818 n=6)
Query/a_2000,_instant_query/Mimir-8                                                    2.719Mi ±  1%    2.722Mi ±  0%       ~ (p=0.699 n=6)
Query/a_2000,_instant_query/Prometheus-8                                               3.051Mi ±  0%    3.048Mi ±  0%       ~ (p=0.589 n=6)
Query/a_2000,_range_query_with_100_steps/Mimir-8                                       3.830Mi ±  1%    3.842Mi ±  0%       ~ (p=0.699 n=6)
Query/a_2000,_range_query_with_100_steps/Prometheus-8                                  4.008Mi ±  0%    4.010Mi ±  0%       ~ (p=0.937 n=6)
Query/a_2000,_range_query_with_1000_steps/Mimir-8                                      12.46Mi ±  0%    12.50Mi ±  0%       ~ (p=0.240 n=6)
Query/a_2000,_range_query_with_1000_steps/Prometheus-8                                 12.62Mi ±  0%    12.62Mi ±  1%       ~ (p=0.699 n=6)
Query/nh_1,_instant_query/Mimir-8                                                      33.91Ki ±  0%    33.91Ki ±  0%       ~ (p=0.916 n=6)
Query/nh_1,_instant_query/Prometheus-8                                                 35.39Ki ±  1%    35.42Ki ±  0%       ~ (p=0.288 n=6)
Query/nh_1,_range_query_with_100_steps/Mimir-8                                         85.23Ki ±  0%    85.26Ki ±  0%       ~ (p=0.818 n=6)
Query/nh_1,_range_query_with_100_steps/Prometheus-8                                    87.14Ki ±  0%    87.16Ki ±  0%       ~ (p=0.589 n=6)
Query/nh_1,_range_query_with_1000_steps/Mimir-8                                        516.3Ki ±  0%    516.3Ki ±  0%       ~ (p=1.000 n=6)
Query/nh_1,_range_query_with_1000_steps/Prometheus-8                                   517.6Ki ±  0%    517.5Ki ±  0%       ~ (p=0.937 n=6)
Query/nh_100,_instant_query/Mimir-8                                                    924.9Ki ±  0%    924.7Ki ±  0%       ~ (p=0.937 n=6)
Query/nh_100,_instant_query/Prometheus-8                                               941.0Ki ±  0%    942.0Ki ±  0%       ~ (p=0.240 n=6)
Query/nh_100,_range_query_with_100_steps/Mimir-8                                       5.565Mi ±  0%    5.567Mi ±  0%  +0.04% (p=0.026 n=6)
Query/nh_100,_range_query_with_100_steps/Prometheus-8                                  5.568Mi ±  0%    5.567Mi ±  0%       ~ (p=0.180 n=6)
Query/nh_100,_range_query_with_1000_steps/Mimir-8                                      47.55Mi ±  0%    47.56Mi ±  0%       ~ (p=0.937 n=6)
Query/nh_100,_range_query_with_1000_steps/Prometheus-8                                 47.54Mi ±  0%    47.56Mi ±  0%       ~ (p=0.310 n=6)
Query/nh_2000,_instant_query/Mimir-8                                                   17.51Mi ±  0%    17.52Mi ±  0%       ~ (p=0.699 n=6)
Query/nh_2000,_instant_query/Prometheus-8                                              17.84Mi ±  0%    17.82Mi ±  0%  -0.12% (p=0.009 n=6)
Query/nh_2000,_range_query_with_100_steps/Mimir-8                                      111.9Mi ±  0%    112.0Mi ±  0%       ~ (p=0.394 n=6)
Query/nh_2000,_range_query_with_100_steps/Prometheus-8                                 111.7Mi ±  0%    112.0Mi ±  0%       ~ (p=0.093 n=6)
Query/nh_2000,_range_query_with_1000_steps/Mimir-8                                     981.1Mi ±  0%    981.2Mi ±  0%       ~ (p=0.699 n=6)
Query/nh_2000,_range_query_with_1000_steps/Prometheus-8                                977.1Mi ±  0%    976.9Mi ±  0%       ~ (p=0.180 n=6)
Query/a_1[1m],_instant_query/Mimir-8                                                   19.01Ki ±  0%    19.08Ki ±  0%  +0.36% (p=0.002 n=6)
Query/a_1[1m],_instant_query/Prometheus-8                                              21.88Ki ±  0%    21.87Ki ±  0%       ~ (p=0.485 n=6)
Query/a_100[1m],_instant_query/Mimir-8                                                 114.3Ki ±  0%    114.4Ki ±  0%       ~ (p=0.093 n=6)
Query/a_100[1m],_instant_query/Prometheus-8                                            139.6Ki ±  0%    139.8Ki ±  0%       ~ (p=0.699 n=6)
Query/a_2000[1m],_instant_query/Mimir-8                                                1.941Mi ±  0%    1.943Mi ±  0%       ~ (p=0.589 n=6)
Query/a_2000[1m],_instant_query/Prometheus-8                                           2.356Mi ±  0%    2.354Mi ±  0%       ~ (p=0.240 n=6)
Query/rate(a_1[1m]),_instant_query/Mimir-8                                             19.24Ki ±  0%    19.34Ki ±  0%  +0.52% (p=0.002 n=6)
Query/rate(a_1[1m]),_instant_query/Prometheus-8                                        22.73Ki ±  0%    22.70Ki ±  0%       ~ (p=0.589 n=6)
Query/rate(a_1[1m]),_range_query_with_100_steps/Mimir-8                                19.67Ki ±  0%    19.76Ki ±  0%  +0.42% (p=0.002 n=6)
Query/rate(a_1[1m]),_range_query_with_100_steps/Prometheus-8                           23.68Ki ±  0%    23.65Ki ±  0%       ~ (p=0.331 n=6)
Query/rate(a_1[1m]),_range_query_with_1000_steps/Mimir-8                               24.51Ki ±  0%    24.59Ki ±  0%  +0.32% (p=0.026 n=6)
Query/rate(a_1[1m]),_range_query_with_1000_steps/Prometheus-8                          28.62Ki ±  0%    28.63Ki ±  0%       ~ (p=1.000 n=6)
Query/rate(a_100[1m]),_instant_query/Mimir-8                                           114.0Ki ±  0%    114.0Ki ±  0%       ~ (p=0.394 n=6)
Query/rate(a_100[1m]),_instant_query/Prometheus-8                                      154.1Ki ±  0%    154.1Ki ±  0%       ~ (p=0.937 n=6)
Query/rate(a_100[1m]),_range_query_with_100_steps/Mimir-8                              158.1Ki ±  0%    158.4Ki ±  0%       ~ (p=0.132 n=6)
Query/rate(a_100[1m]),_range_query_with_100_steps/Prometheus-8                         193.5Ki ±  0%    193.5Ki ±  0%       ~ (p=0.818 n=6)
Query/rate(a_100[1m]),_range_query_with_1000_steps/Mimir-8                             633.3Ki ±  0%    636.1Ki ±  1%       ~ (p=0.132 n=6)
Query/rate(a_100[1m]),_range_query_with_1000_steps/Prometheus-8                        671.6Ki ±  1%    672.9Ki ±  1%       ~ (p=1.000 n=6)
Query/rate(a_2000[1m]),_instant_query/Mimir-8                                          1.862Mi ±  1%    1.865Mi ±  0%       ~ (p=0.310 n=6)
Query/rate(a_2000[1m]),_instant_query/Prometheus-8                                     2.644Mi ±  0%    2.645Mi ±  0%       ~ (p=0.310 n=6)
Query/rate(a_2000[1m]),_range_query_with_100_steps/Mimir-8                             2.802Mi ±  1%    2.794Mi ±  0%       ~ (p=0.485 n=6)
Query/rate(a_2000[1m]),_range_query_with_100_steps/Prometheus-8                        3.421Mi ±  3%    3.420Mi ±  0%       ~ (p=0.937 n=6)
Query/rate(a_2000[1m]),_range_query_with_1000_steps/Mimir-8                            11.89Mi ±  1%    11.91Mi ±  1%       ~ (p=0.699 n=6)
Query/rate(a_2000[1m]),_range_query_with_1000_steps/Prometheus-8                       12.49Mi ±  1%    12.49Mi ±  0%       ~ (p=1.000 n=6)
Query/rate(a_1[1m]),_range_query_with_10000_steps/Mimir-8                              76.76Ki ±  3%    76.55Ki ±  4%       ~ (p=0.937 n=6)
Query/rate(a_1[1m]),_range_query_with_10000_steps/Prometheus-8                         88.62Ki ±  9%    87.79Ki ± 10%       ~ (p=0.818 n=6)
Query/rate(a_100[1m]),_range_query_with_10000_steps/Mimir-8                            5.513Mi ±  2%    5.622Mi ±  1%  +1.99% (p=0.026 n=6)
Query/rate(a_100[1m]),_range_query_with_10000_steps/Prometheus-8                       5.623Mi ±  3%    5.562Mi ±  3%       ~ (p=0.818 n=6)
Query/rate(a_2000[1m]),_range_query_with_10000_steps/Mimir-8                           610.3Mi ±  0%    610.0Mi ±  0%       ~ (p=0.394 n=6)
Query/rate(a_2000[1m]),_range_query_with_10000_steps/Prometheus-8                      1.054Gi ±  0%    1.054Gi ±  0%       ~ (p=0.485 n=6)
Query/rate(a_1[1d]),_instant_query/Mimir-8                                             76.24Ki ±  5%    71.35Ki ±  8%  -6.41% (p=0.041 n=6)
Query/rate(a_1[1d]),_instant_query/Prometheus-8                                        642.9Ki ±  2%    636.6Ki ±  3%       ~ (p=0.937 n=6)
Query/rate(a_1[1d]),_range_query_with_100_steps/Mimir-8                                72.68Ki ± 16%    77.84Ki ±  6%       ~ (p=0.180 n=6)
Query/rate(a_1[1d]),_range_query_with_100_steps/Prometheus-8                           640.3Ki ±  5%    647.9Ki ±  2%       ~ (p=1.000 n=6)
Query/rate(a_1[1d]),_range_query_with_1000_steps/Mimir-8                               69.05Ki ± 16%    73.25Ki ± 39%       ~ (p=0.394 n=6)
Query/rate(a_1[1d]),_range_query_with_1000_steps/Prometheus-8                          638.2Ki ±  5%    641.0Ki ±  6%       ~ (p=0.818 n=6)
Query/rate(a_100[1d]),_instant_query/Mimir-8                                           5.089Mi ±  3%    5.099Mi ±  3%       ~ (p=1.000 n=6)
Query/rate(a_100[1d]),_instant_query/Prometheus-8                                      5.660Mi ±  2%    5.539Mi ±  3%       ~ (p=0.310 n=6)
Query/rate(a_100[1d]),_range_query_with_100_steps/Mimir-8                              5.105Mi ±  3%    5.150Mi ±  2%       ~ (p=0.818 n=6)
Query/rate(a_100[1d]),_range_query_with_100_steps/Prometheus-8                         5.667Mi ±  4%    5.676Mi ±  3%       ~ (p=0.699 n=6)
Query/rate(a_100[1d]),_range_query_with_1000_steps/Mimir-8                             5.899Mi ±  6%    5.915Mi ±  8%       ~ (p=0.937 n=6)
Query/rate(a_100[1d]),_range_query_with_1000_steps/Prometheus-8                        8.367Mi ±  2%    8.421Mi ±  1%       ~ (p=0.310 n=6)
Query/rate(a_2000[1d]),_instant_query/Mimir-8                                          96.21Mi ±  1%    96.28Mi ±  1%       ~ (p=0.699 n=6)
Query/rate(a_2000[1d]),_instant_query/Prometheus-8                                     97.41Mi ±  1%    97.44Mi ±  1%       ~ (p=0.818 n=6)
Query/rate(a_2000[1d]),_range_query_with_100_steps/Mimir-8                             101.2Mi ±  0%    100.8Mi ±  1%       ~ (p=0.132 n=6)
Query/rate(a_2000[1d]),_range_query_with_100_steps/Prometheus-8                        101.1Mi ±  1%    101.4Mi ±  1%       ~ (p=0.818 n=6)
Query/rate(a_2000[1d]),_range_query_with_1000_steps/Mimir-8                            135.3Mi ±  0%    135.5Mi ±  0%       ~ (p=0.310 n=6)
Query/rate(a_2000[1d]),_range_query_with_1000_steps/Prometheus-8                       136.3Mi ±  0%    136.4Mi ±  1%       ~ (p=0.485 n=6)
Query/a_1_-_b_1,_instant_query/Mimir-8                                                 37.97Ki ±  0%    37.95Ki ±  0%       ~ (p=0.485 n=6)
Query/a_1_-_b_1,_instant_query/Prometheus-8                                            42.52Ki ±  0%    42.52Ki ±  0%       ~ (p=0.937 n=6)
Query/a_1_-_b_1,_range_query_with_100_steps/Mimir-8                                    39.05Ki ±  0%    39.07Ki ±  0%       ~ (p=0.699 n=6)
Query/a_1_-_b_1,_range_query_with_100_steps/Prometheus-8                               48.79Ki ±  0%    48.80Ki ±  0%       ~ (p=0.855 n=6)
Query/a_1_-_b_1,_range_query_with_1000_steps/Mimir-8                                   48.44Ki ±  0%    48.43Ki ±  0%       ~ (p=0.818 n=6)
Query/a_1_-_b_1,_range_query_with_1000_steps/Prometheus-8                              100.9Ki ±  0%    100.8Ki ±  0%       ~ (p=1.000 n=6)
Query/a_100_-_b_100,_instant_query/Mimir-8                                             338.2Ki ±  0%    338.1Ki ±  0%       ~ (p=0.485 n=6)
Query/a_100_-_b_100,_instant_query/Prometheus-8                                        437.8Ki ±  0%    437.8Ki ±  0%       ~ (p=1.000 n=6)
Query/a_100_-_b_100,_range_query_with_100_steps/Mimir-8                                451.7Ki ±  0%    452.0Ki ±  0%       ~ (p=0.461 n=6)
Query/a_100_-_b_100,_range_query_with_100_steps/Prometheus-8                           643.1Ki ±  0%    643.5Ki ±  0%       ~ (p=0.485 n=6)
Query/a_100_-_b_100,_range_query_with_1000_steps/Mimir-8                               1.322Mi ±  0%    1.322Mi ±  0%       ~ (p=0.818 n=6)
Query/a_100_-_b_100,_range_query_with_1000_steps/Prometheus-8                          2.352Mi ±  1%    2.350Mi ±  0%       ~ (p=0.485 n=6)
Query/a_2000_-_b_2000,_instant_query/Mimir-8                                           6.017Mi ±  0%    6.021Mi ±  0%       ~ (p=0.818 n=6)
Query/a_2000_-_b_2000,_instant_query/Prometheus-8                                      8.271Mi ±  0%    8.250Mi ±  0%       ~ (p=0.132 n=6)
Query/a_2000_-_b_2000,_range_query_with_100_steps/Mimir-8                              8.227Mi ±  0%    8.218Mi ±  1%       ~ (p=1.000 n=6)
Query/a_2000_-_b_2000,_range_query_with_100_steps/Prometheus-8                         10.36Mi ±  1%    10.45Mi ± 11%       ~ (p=0.310 n=6)
Query/a_2000_-_b_2000,_range_query_with_1000_steps/Mimir-8                             25.74Mi ±  1%    25.48Mi ±  1%  -1.02% (p=0.041 n=6)
Query/a_2000_-_b_2000,_range_query_with_1000_steps/Prometheus-8                        123.1Mi ±  0%    123.0Mi ±  0%       ~ (p=1.000 n=6)
Query/a_1_-_b_1,_range_query_with_10000_steps/Mimir-8                                  155.7Ki ±  4%    152.2Ki ±  4%       ~ (p=0.310 n=6)
Query/a_1_-_b_1,_range_query_with_10000_steps/Prometheus-8                             680.1Ki ±  3%    684.4Ki ±  5%       ~ (p=0.699 n=6)
Query/a_100_-_b_100,_range_query_with_10000_steps/Mimir-8                              10.92Mi ±  1%    10.99Mi ±  1%       ~ (p=0.132 n=6)
Query/a_100_-_b_100,_range_query_with_10000_steps/Prometheus-8                         21.08Mi ±  1%    20.86Mi ±  1%       ~ (p=0.180 n=6)
Query/a_2000_-_b_2000,_range_query_with_10000_steps/Mimir-8                            722.3Mi ±  0%    721.2Mi ±  0%       ~ (p=0.394 n=6)
Query/a_2000_-_b_2000,_range_query_with_10000_steps/Prometheus-8                       3.056Gi ±  0%    3.055Gi ±  0%       ~ (p=0.132 n=6)
Query/a_100{l=~"[13579]."}_-_b_100,_instant_query/Mimir-8                              258.4Ki ±  0%    257.8Ki ±  0%  -0.21% (p=0.009 n=6)
Query/a_100{l=~"[13579]."}_-_b_100,_instant_query/Prometheus-8                         340.6Ki ±  0%    340.5Ki ±  0%       ~ (p=1.000 n=6)
Query/a_100{l=~"[13579]."}_-_b_100,_range_query_with_100_steps/Mimir-8                 340.4Ki ±  0%    340.1Ki ±  0%       ~ (p=0.394 n=6)
Query/a_100{l=~"[13579]."}_-_b_100,_range_query_with_100_steps/Prometheus-8            508.4Ki ±  0%    508.4Ki ±  0%       ~ (p=0.937 n=6)
Query/a_100{l=~"[13579]."}_-_b_100,_range_query_with_1000_steps/Mimir-8               1017.3Ki ±  0%   1021.3Ki ±  1%       ~ (p=0.065 n=6)
Query/a_100{l=~"[13579]."}_-_b_100,_range_query_with_1000_steps/Prometheus-8           1.944Mi ±  0%    1.949Mi ±  0%       ~ (p=0.394 n=6)
Query/a_2000{l=~"1..."}_-_b_2000,_instant_query/Mimir-8                                4.059Mi ±  0%    4.052Mi ±  0%       ~ (p=0.132 n=6)
Query/a_2000{l=~"1..."}_-_b_2000,_instant_query/Prometheus-8                           6.221Mi ±  0%    6.219Mi ±  0%       ~ (p=0.937 n=6)
Query/a_2000{l=~"1..."}_-_b_2000,_range_query_with_100_steps/Mimir-8                   5.509Mi ±  1%    5.505Mi ±  1%       ~ (p=0.394 n=6)
Query/a_2000{l=~"1..."}_-_b_2000,_range_query_with_100_steps/Prometheus-8              7.760Mi ±  0%    7.758Mi ±  0%       ~ (p=0.699 n=6)
Query/a_2000{l=~"1..."}_-_b_2000,_range_query_with_1000_steps/Mimir-8                  16.82Mi ±  1%    16.84Mi ±  0%       ~ (p=0.589 n=6)
Query/a_2000{l=~"1..."}_-_b_2000,_range_query_with_1000_steps/Prometheus-8             21.17Mi ±  1%    21.34Mi ±  2%       ~ (p=0.394 n=6)
Query/a_2000{l="1234"}_-_b_2000,_instant_query/Mimir-8                                 1.994Mi ±  1%    1.992Mi ±  3%       ~ (p=0.699 n=6)
Query/a_2000{l="1234"}_-_b_2000,_instant_query/Prometheus-8                            4.291Mi ±  0%    4.293Mi ±  0%       ~ (p=0.240 n=6)
Query/a_2000{l="1234"}_-_b_2000,_range_query_with_100_steps/Mimir-8                    2.467Mi ±  6%    2.551Mi ±  4%       ~ (p=0.589 n=6)
Query/a_2000{l="1234"}_-_b_2000,_range_query_with_100_steps/Prometheus-8               5.362Mi ±  1%    5.362Mi ±  1%       ~ (p=0.699 n=6)
Query/a_2000{l="1234"}_-_b_2000,_range_query_with_1000_steps/Mimir-8                   5.890Mi ±  7%    5.802Mi ±  6%       ~ (p=0.240 n=6)
Query/a_2000{l="1234"}_-_b_2000,_range_query_with_1000_steps/Prometheus-8              14.01Mi ±  1%    14.02Mi ±  1%       ~ (p=0.394 n=6)
Query/a_2000_-_b_2000{l="1234"},_instant_query/Mimir-8                                 2.071Mi ±  1%    2.055Mi ±  1%       ~ (p=0.065 n=6)
Query/a_2000_-_b_2000{l="1234"},_instant_query/Prometheus-8                            3.662Mi ±  0%    3.663Mi ±  0%       ~ (p=0.589 n=6)
Query/a_2000_-_b_2000{l="1234"},_range_query_with_100_steps/Mimir-8                    2.584Mi ±  5%    2.481Mi ±  8%       ~ (p=0.240 n=6)
Query/a_2000_-_b_2000{l="1234"},_range_query_with_100_steps/Prometheus-8               4.739Mi ±  0%    4.735Mi ±  0%       ~ (p=0.818 n=6)
Query/a_2000_-_b_2000{l="1234"},_range_query_with_1000_steps/Mimir-8                   5.918Mi ±  4%    5.910Mi ±  3%       ~ (p=0.937 n=6)
Query/a_2000_-_b_2000{l="1234"},_range_query_with_1000_steps/Prometheus-8              13.64Mi ±  1%    13.59Mi ±  1%       ~ (p=0.394 n=6)
Query/sum(a_1),_instant_query/Mimir-8                                                  19.94Ki ±  0%    19.94Ki ±  0%       ~ (p=0.818 n=6)
Query/sum(a_1),_instant_query/Prometheus-8                                             21.92Ki ±  1%    21.91Ki ±  0%       ~ (p=1.000 n=6)
Query/sum(a_1),_range_query_with_100_steps/Mimir-8                                     20.52Ki ±  0%    20.51Ki ±  0%       ~ (p=0.786 n=6)
Query/sum(a_1),_range_query_with_100_steps/Prometheus-8                                23.02Ki ±  0%    22.98Ki ±  2%       ~ (p=0.180 n=6)
Query/sum(a_1),_range_query_with_1000_steps/Mimir-8                                    25.28Ki ±  1%    25.23Ki ±  1%       ~ (p=0.310 n=6)
Query/sum(a_1),_range_query_with_1000_steps/Prometheus-8                               27.77Ki ±  0%    27.77Ki ±  0%       ~ (p=0.974 n=6)
Query/sum(a_100),_instant_query/Mimir-8                                                158.1Ki ±  0%    158.2Ki ±  0%       ~ (p=0.240 n=6)
Query/sum(a_100),_instant_query/Prometheus-8                                           177.8Ki ±  0%    177.8Ki ±  0%       ~ (p=0.818 n=6)
Query/sum(a_100),_range_query_with_100_steps/Mimir-8                                   214.3Ki ±  0%    214.4Ki ±  0%       ~ (p=0.143 n=6)
Query/sum(a_100),_range_query_with_100_steps/Prometheus-8                              233.6Ki ±  0%    234.4Ki ±  0%  +0.35% (p=0.026 n=6)
Query/sum(a_100),_range_query_with_1000_steps/Mimir-8                                  665.0Ki ±  1%    665.4Ki ±  1%       ~ (p=0.589 n=6)
Query/sum(a_100),_range_query_with_1000_steps/Prometheus-8                             681.7Ki ±  0%    681.2Ki ±  1%       ~ (p=0.937 n=6)
Query/sum(a_2000),_instant_query/Mimir-8                                               2.714Mi ±  1%    2.717Mi ±  0%       ~ (p=0.180 n=6)
Query/sum(a_2000),_instant_query/Prometheus-8                                          3.114Mi ±  0%    3.112Mi ±  0%       ~ (p=0.699 n=6)
Query/sum(a_2000),_range_query_with_100_steps/Mimir-8                                  3.758Mi ±  0%    3.755Mi ±  0%       ~ (p=0.485 n=6)
Query/sum(a_2000),_range_query_with_100_steps/Prometheus-8                             4.134Mi ±  1%    4.137Mi ±  3%       ~ (p=0.818 n=6)
Query/sum(a_2000),_range_query_with_1000_steps/Mimir-8                                 12.56Mi ±  1%    12.53Mi ±  1%       ~ (p=0.818 n=6)
Query/sum(a_2000),_range_query_with_1000_steps/Prometheus-8                            12.81Mi ±  1%    12.76Mi ±  0%       ~ (p=0.132 n=6)
Query/sum_by_(l)(h_1),_instant_query/Mimir-8                                           26.92Ki ±  0%    26.92Ki ±  0%       ~ (p=0.699 n=6)
Query/sum_by_(l)(h_1),_instant_query/Prometheus-8                                      29.91Ki ±  0%    29.86Ki ±  0%       ~ (p=0.240 n=6)
Query/sum_by_(l)(h_1),_range_query_with_100_steps/Mimir-8                              30.16Ki ±  0%    30.19Ki ±  0%       ~ (p=0.093 n=6)
Query/sum_by_(l)(h_1),_range_query_with_100_steps/Prometheus-8                         33.65Ki ±  0%    33.63Ki ±  0%       ~ (p=1.000 n=6)
Query/sum_by_(l)(h_1),_range_query_with_1000_steps/Mimir-8                             58.01Ki ±  0%    58.04Ki ±  0%       ~ (p=0.818 n=6)
Query/sum_by_(l)(h_1),_range_query_with_1000_steps/Prometheus-8                        61.20Ki ±  0%    61.17Ki ±  0%       ~ (p=0.485 n=6)
Query/sum_by_(l)(h_100),_instant_query/Mimir-8                                         941.9Ki ±  0%    941.8Ki ±  0%       ~ (p=0.937 n=6)
Query/sum_by_(l)(h_100),_instant_query/Prometheus-8                                    1.072Mi ±  0%    1.072Mi ±  0%       ~ (p=0.699 n=6)
Query/sum_by_(l)(h_100),_range_query_with_100_steps/Mimir-8                            1.239Mi ±  0%    1.238Mi ±  0%       ~ (p=0.485 n=6)
Query/sum_by_(l)(h_100),_range_query_with_100_steps/Prometheus-8                       1.381Mi ±  0%    1.379Mi ±  0%       ~ (p=0.310 n=6)
Query/sum_by_(l)(h_100),_range_query_with_1000_steps/Mimir-8                           3.891Mi ±  1%    3.887Mi ±  1%       ~ (p=0.937 n=6)
Query/sum_by_(l)(h_100),_range_query_with_1000_steps/Prometheus-8                      4.040Mi ±  0%    4.041Mi ±  1%       ~ (p=0.937 n=6)
Query/sum_by_(l)(h_2000),_instant_query/Mimir-8                                        19.82Mi ±  0%    19.86Mi ±  0%       ~ (p=0.180 n=6)
Query/sum_by_(l)(h_2000),_instant_query/Prometheus-8                                   22.80Mi ±  0%    22.78Mi ±  0%       ~ (p=0.589 n=6)
Query/sum_by_(l)(h_2000),_range_query_with_100_steps/Mimir-8                           27.45Mi ±  2%    27.13Mi ±  3%       ~ (p=0.699 n=6)
Query/sum_by_(l)(h_2000),_range_query_with_100_steps/Prometheus-8                      28.62Mi ±  0%    28.61Mi ±  0%       ~ (p=0.937 n=6)
Query/sum_by_(l)(h_2000),_range_query_with_1000_steps/Mimir-8                          111.1Mi ±  0%    111.1Mi ±  0%       ~ (p=0.699 n=6)
Query/sum_by_(l)(h_2000),_range_query_with_1000_steps/Prometheus-8                     300.6Mi ±  0%    300.7Mi ±  0%       ~ (p=1.000 n=6)
Query/sum_by_(le)(h_1),_instant_query/Mimir-8                                          27.08Ki ±  0%    27.07Ki ±  0%       ~ (p=0.420 n=6)
Query/sum_by_(le)(h_1),_instant_query/Prometheus-8                                     31.39Ki ±  0%    31.38Ki ±  0%       ~ (p=0.937 n=6)
Query/sum_by_(le)(h_1),_range_query_with_100_steps/Mimir-8                             30.38Ki ±  0%    30.36Ki ±  0%       ~ (p=0.589 n=6)
Query/sum_by_(le)(h_1),_range_query_with_100_steps/Prometheus-8                        34.98Ki ±  0%    34.99Ki ±  0%       ~ (p=0.905 n=6)
Query/sum_by_(le)(h_1),_range_query_with_1000_steps/Mimir-8                            58.02Ki ±  0%    58.21Ki ±  1%       ~ (p=0.394 n=6)
Query/sum_by_(le)(h_1),_range_query_with_1000_steps/Prometheus-8                       62.60Ki ±  0%    62.61Ki ±  0%       ~ (p=0.699 n=6)
Query/sum_by_(le)(h_100),_instant_query/Mimir-8                                        924.1Ki ±  0%    924.6Ki ±  0%       ~ (p=0.394 n=6)
Query/sum_by_(le)(h_100),_instant_query/Prometheus-8                                   1.035Mi ±  0%    1.034Mi ±  0%       ~ (p=0.937 n=6)
Query/sum_by_(le)(h_100),_range_query_with_100_steps/Mimir-8                           1.218Mi ±  0%    1.217Mi ±  0%       ~ (p=0.589 n=6)
Query/sum_by_(le)(h_100),_range_query_with_100_steps/Prometheus-8                      1.348Mi ±  0%    1.349Mi ±  0%       ~ (p=0.485 n=6)
Query/sum_by_(le)(h_100),_range_query_with_1000_steps/Mimir-8                          3.864Mi ±  0%    3.871Mi ±  1%       ~ (p=0.240 n=6)
Query/sum_by_(le)(h_100),_range_query_with_1000_steps/Prometheus-8                     3.963Mi ±  1%    3.970Mi ±  1%       ~ (p=0.818 n=6)
Query/sum_by_(le)(h_2000),_instant_query/Mimir-8                                       19.28Mi ±  0%    19.26Mi ±  1%       ~ (p=0.310 n=6)
Query/sum_by_(le)(h_2000),_instant_query/Prometheus-8                                  21.93Mi ±  0%    21.95Mi ±  0%       ~ (p=0.699 n=6)
Query/sum_by_(le)(h_2000),_range_query_with_100_steps/Mimir-8                          25.59Mi ±  1%    25.49Mi ±  1%  -0.41% (p=0.041 n=6)
Query/sum_by_(le)(h_2000),_range_query_with_100_steps/Prometheus-8                     27.88Mi ±  0%    27.90Mi ±  0%       ~ (p=0.589 n=6)
Query/sum_by_(le)(h_2000),_range_query_with_1000_steps/Mimir-8                         78.96Mi ±  1%    78.86Mi ±  0%       ~ (p=0.699 n=6)
Query/sum_by_(le)(h_2000),_range_query_with_1000_steps/Prometheus-8                    268.8Mi ±  0%    268.7Mi ±  0%       ~ (p=0.589 n=6)
Query/rate(a_1[1m])_+_rate(b_1[1m]),_instant_query/Mimir-8                             37.71Ki ±  0%    37.95Ki ±  0%  +0.63% (p=0.002 n=6)
Query/rate(a_1[1m])_+_rate(b_1[1m]),_instant_query/Prometheus-8                        46.31Ki ±  0%    46.28Ki ±  0%       ~ (p=0.180 n=6)
Query/rate(a_1[1m])_+_rate(b_1[1m]),_range_query_with_100_steps/Mimir-8                38.56Ki ±  0%    38.77Ki ±  0%  +0.55% (p=0.002 n=6)
Query/rate(a_1[1m])_+_rate(b_1[1m]),_range_query_with_100_steps/Prometheus-8           52.33Ki ±  0%    52.36Ki ±  0%       ~ (p=0.781 n=6)
Query/rate(a_1[1m])_+_rate(b_1[1m]),_range_query_with_1000_steps/Mimir-8               48.15Ki ±  0%    48.30Ki ±  1%  +0.31% (p=0.009 n=6)
Query/rate(a_1[1m])_+_rate(b_1[1m]),_range_query_with_1000_steps/Prometheus-8          105.0Ki ±  0%    104.8Ki ±  0%       ~ (p=0.180 n=6)
Query/rate(a_100[1m])_+_rate(b_100[1m]),_instant_query/Mimir-8                         247.4Ki ±  0%    247.8Ki ±  0%       ~ (p=0.288 n=6)
Query/rate(a_100[1m])_+_rate(b_100[1m]),_instant_query/Prometheus-8                    394.1Ki ±  0%    394.1Ki ±  0%       ~ (p=1.000 n=6)
Query/rate(a_100[1m])_+_rate(b_100[1m]),_range_query_with_100_steps/Mimir-8            335.2Ki ±  0%    335.7Ki ±  0%       ~ (p=0.818 n=6)
Query/rate(a_100[1m])_+_rate(b_100[1m]),_range_query_with_100_steps/Prometheus-8       574.0Ki ±  1%    574.5Ki ±  1%       ~ (p=0.699 n=6)
Query/rate(a_100[1m])_+_rate(b_100[1m]),_range_query_with_1000_steps/Mimir-8           1.261Mi ±  1%    1.262Mi ±  0%       ~ (p=0.818 n=6)
Query/rate(a_100[1m])_+_rate(b_100[1m]),_range_query_with_1000_steps/Prometheus-8      2.335Mi ±  1%    2.334Mi ±  0%       ~ (p=0.485 n=6)
Query/rate(a_2000[1m])_+_rate(b_2000[1m]),_instant_query/Mimir-8                       4.222Mi ±  0%    4.226Mi ±  0%       ~ (p=0.589 n=6)
Query/rate(a_2000[1m])_+_rate(b_2000[1m]),_instant_query/Prometheus-8                  7.354Mi ±  0%    7.355Mi ±  0%       ~ (p=0.240 n=6)
Query/rate(a_2000[1m])_+_rate(b_2000[1m]),_range_query_with_100_steps/Mimir-8          6.060Mi ±  1%    6.062Mi ±  1%       ~ (p=0.589 n=6)
Query/rate(a_2000[1m])_+_rate(b_2000[1m]),_range_query_with_100_steps/Prometheus-8     9.025Mi ±  0%    9.021Mi ±  0%       ~ (p=0.818 n=6)
Query/rate(a_2000[1m])_+_rate(b_2000[1m]),_range_query_with_1000_steps/Mimir-8         25.11Mi ±  4%    25.37Mi ±  2%       ~ (p=0.699 n=6)
Query/rate(a_2000[1m])_+_rate(b_2000[1m]),_range_query_with_1000_steps/Prometheus-8    122.9Mi ±  0%    122.8Mi ±  0%       ~ (p=0.818 n=6)
Query/sum(a_1_+_b_1),_instant_query/Mimir-8                                            38.63Ki ±  0%    38.60Ki ±  0%       ~ (p=0.485 n=6)
Query/sum(a_1_+_b_1),_instant_query/Prometheus-8                                       43.88Ki ±  0%    43.92Ki ±  0%       ~ (p=0.240 n=6)
Query/sum(a_1_+_b_1),_range_query_with_100_steps/Mimir-8                               39.73Ki ±  0%    39.70Ki ±  0%       ~ (p=0.699 n=6)
Query/sum(a_1_+_b_1),_range_query_with_100_steps/Prometheus-8                          50.17Ki ±  0%    50.17Ki ±  0%       ~ (p=1.000 n=6)
Query/sum(a_1_+_b_1),_range_query_with_1000_steps/Mimir-8                              49.22Ki ±  1%    49.15Ki ±  1%       ~ (p=0.485 n=6)
Query/sum(a_1_+_b_1),_range_query_with_1000_steps/Prometheus-8                         102.2Ki ±  0%    102.2Ki ±  0%       ~ (p=0.853 n=6)
Query/sum(a_100_+_b_100),_instant_query/Mimir-8                                        339.9Ki ±  0%    339.7Ki ±  0%       ~ (p=0.093 n=6)
Query/sum(a_100_+_b_100),_instant_query/Prometheus-8                                   442.4Ki ±  0%    442.7Ki ±  0%       ~ (p=0.485 n=6)
Query/sum(a_100_+_b_100),_range_query_with_100_steps/Mimir-8                           451.7Ki ±  0%    451.2Ki ±  0%       ~ (p=0.310 n=6)
Query/sum(a_100_+_b_100),_range_query_with_100_steps/Prometheus-8                      650.8Ki ±  0%    651.6Ki ±  0%       ~ (p=1.000 n=6)
Query/sum(a_100_+_b_100),_range_query_with_1000_steps/Mimir-8                          1.325Mi ±  1%    1.323Mi ±  0%       ~ (p=0.818 n=6)
Query/sum(a_100_+_b_100),_range_query_with_1000_steps/Prometheus-8                     2.358Mi ±  0%    2.368Mi ±  1%  +0.42% (p=0.004 n=6)
Query/sum(a_2000_+_b_2000),_instant_query/Mimir-8                                      5.998Mi ±  1%    5.993Mi ±  0%       ~ (p=0.485 n=6)
Query/sum(a_2000_+_b_2000),_instant_query/Prometheus-8                                 8.331Mi ±  0%    8.336Mi ±  0%       ~ (p=0.699 n=6)
Query/sum(a_2000_+_b_2000),_range_query_with_100_steps/Mimir-8                         8.134Mi ±  0%    8.129Mi ±  0%       ~ (p=0.699 n=6)
Query/sum(a_2000_+_b_2000),_range_query_with_100_steps/Prometheus-8                    10.55Mi ± 17%    10.52Mi ±  1%       ~ (p=0.240 n=6)
Query/sum(a_2000_+_b_2000),_range_query_with_1000_steps/Mimir-8                        25.47Mi ±  2%    25.50Mi ±  1%       ~ (p=0.699 n=6)
Query/sum(a_2000_+_b_2000),_range_query_with_1000_steps/Prometheus-8                   123.5Mi ±  1%    123.3Mi ±  1%       ~ (p=0.394 n=6)
Query/sum_by_(le)(rate(h_1[1m])),_instant_query/Mimir-8                                24.78Ki ±  0%    24.86Ki ±  0%  +0.29% (p=0.002 n=6)
Query/sum_by_(le)(rate(h_1[1m])),_instant_query/Prometheus-8                           32.74Ki ±  0%    32.73Ki ±  1%       ~ (p=0.855 n=6)
Query/sum_by_(le)(rate(h_1[1m])),_range_query_with_100_steps/Mimir-8                   27.35Ki ±  0%    27.45Ki ±  0%  +0.37% (p=0.009 n=6)
Query/sum_by_(le)(rate(h_1[1m])),_range_query_with_100_steps/Prometheus-8              35.23Ki ±  0%    35.25Ki ±  0%       ~ (p=0.589 n=6)
Query/sum_by_(le)(rate(h_1[1m])),_range_query_with_1000_steps/Mimir-8                  56.09Ki ±  0%    56.29Ki ±  1%       ~ (p=0.065 n=6)
Query/sum_by_(le)(rate(h_1[1m])),_range_query_with_1000_steps/Prometheus-8             63.98Ki ±  1%    64.04Ki ±  0%       ~ (p=0.485 n=6)
Query/sum_by_(le)(rate(h_100[1m])),_instant_query/Mimir-8                              664.1Ki ±  0%    664.4Ki ±  0%       ~ (p=0.310 n=6)
Query/sum_by_(le)(rate(h_100[1m])),_instant_query/Prometheus-8                         935.3Ki ±  0%    934.6Ki ±  0%       ~ (p=0.485 n=6)
Query/sum_by_(le)(rate(h_100[1m])),_range_query_with_100_steps/Mimir-8                 928.2Ki ±  0%    928.6Ki ±  0%       ~ (p=0.699 n=6)
Query/sum_by_(le)(rate(h_100[1m])),_range_query_with_100_steps/Prometheus-8            1.173Mi ±  0%    1.172Mi ±  0%       ~ (p=0.818 n=6)
Query/sum_by_(le)(rate(h_100[1m])),_range_query_with_1000_steps/Mimir-8                3.697Mi ±  0%    3.692Mi ±  0%       ~ (p=0.485 n=6)
Query/sum_by_(le)(rate(h_100[1m])),_range_query_with_1000_steps/Prometheus-8           3.999Mi ±  1%    4.002Mi ±  0%       ~ (p=0.937 n=6)
Query/sum_by_(le)(rate(h_2000[1m])),_instant_query/Mimir-8                             14.02Mi ±  0%    14.04Mi ±  1%       ~ (p=0.818 n=6)
Query/sum_by_(le)(rate(h_2000[1m])),_instant_query/Prometheus-8                        19.38Mi ±  0%    19.38Mi ±  0%       ~ (p=0.818 n=6)
Query/sum_by_(le)(rate(h_2000[1m])),_range_query_with_100_steps/Mimir-8                19.24Mi ±  0%    19.28Mi ±  1%       ~ (p=0.699 n=6)
Query/sum_by_(le)(rate(h_2000[1m])),_range_query_with_100_steps/Prometheus-8           24.30Mi ±  0%    24.30Mi ±  0%       ~ (p=0.937 n=6)
Query/sum_by_(le)(rate(h_2000[1m])),_range_query_with_1000_steps/Mimir-8               75.65Mi ±  0%    75.36Mi ±  1%  -0.39% (p=0.015 n=6)
Query/sum_by_(le)(rate(h_2000[1m])),_range_query_with_1000_steps/Prometheus-8          267.8Mi ±  0%    267.8Mi ±  0%       ~ (p=0.485 n=6)
geomean                                                                                1.196Mi          1.196Mi        +0.03%

                                                                                    │  main.txt   │   mqe-histogram-range-vectors.txt   │
                                                                                    │  allocs/op  │  allocs/op   vs base                │
Query/a_1,_instant_query/Mimir-8                                                       336.0 ± 0%    336.0 ± 0%       ~ (p=1.000 n=6) ¹
Query/a_1,_instant_query/Prometheus-8                                                  365.0 ± 0%    365.0 ± 0%       ~ (p=1.000 n=6) ¹
Query/a_1,_range_query_with_100_steps/Mimir-8                                          341.0 ± 0%    341.0 ± 0%       ~ (p=1.000 n=6) ¹
Query/a_1,_range_query_with_100_steps/Prometheus-8                                     375.0 ± 0%    375.0 ± 0%       ~ (p=1.000 n=6) ¹
Query/a_1,_range_query_with_1000_steps/Mimir-8                                         373.0 ± 0%    373.0 ± 0%       ~ (p=1.000 n=6) ¹
Query/a_1,_range_query_with_1000_steps/Prometheus-8                                    407.0 ± 0%    407.0 ± 0%       ~ (p=1.000 n=6) ¹
Query/a_100,_instant_query/Mimir-8                                                    2.246k ± 0%   2.246k ± 0%       ~ (p=1.000 n=6)
Query/a_100,_instant_query/Prometheus-8                                               2.282k ± 0%   2.283k ± 0%       ~ (p=0.182 n=6)
Query/a_100,_range_query_with_100_steps/Mimir-8                                       2.654k ± 0%   2.654k ± 0%       ~ (p=1.000 n=6)
Query/a_100,_range_query_with_100_steps/Prometheus-8                                  2.694k ± 0%   2.694k ± 0%       ~ (p=1.000 n=6)
Query/a_100,_range_query_with_1000_steps/Mimir-8                                      5.862k ± 0%   5.861k ± 0%       ~ (p=0.634 n=6)
Query/a_100,_range_query_with_1000_steps/Prometheus-8                                 5.899k ± 0%   5.899k ± 0%       ~ (p=0.868 n=6)
Query/a_2000,_instant_query/Mimir-8                                                   38.74k ± 0%   38.74k ± 0%       ~ (p=0.569 n=6)
Query/a_2000,_instant_query/Prometheus-8                                              38.81k ± 0%   38.80k ± 0%       ~ (p=0.143 n=6)
Query/a_2000,_range_query_with_100_steps/Mimir-8                                      46.80k ± 0%   46.80k ± 0%       ~ (p=0.509 n=6)
Query/a_2000,_range_query_with_100_steps/Prometheus-8                                 46.85k ± 0%   46.86k ± 0%       ~ (p=0.900 n=6)
Query/a_2000,_range_query_with_1000_steps/Mimir-8                                     110.9k ± 0%   110.9k ± 0%       ~ (p=0.844 n=6)
Query/a_2000,_range_query_with_1000_steps/Prometheus-8                                110.9k ± 0%   110.9k ± 0%       ~ (p=0.615 n=6)
Query/nh_1,_instant_query/Mimir-8                                                      604.0 ± 0%    604.0 ± 0%       ~ (p=1.000 n=6) ¹
Query/nh_1,_instant_query/Prometheus-8                                                 633.0 ± 0%    633.0 ± 0%       ~ (p=1.000 n=6) ¹
Query/nh_1,_range_query_with_100_steps/Mimir-8                                        1.678k ± 0%   1.678k ± 0%       ~ (p=1.000 n=6)
Query/nh_1,_range_query_with_100_steps/Prometheus-8                                   1.711k ± 0%   1.711k ± 0%       ~ (p=1.000 n=6)
Query/nh_1,_range_query_with_1000_steps/Mimir-8                                       10.75k ± 0%   10.75k ± 0%       ~ (p=0.567 n=6)
Query/nh_1,_range_query_with_1000_steps/Prometheus-8                                  10.79k ± 0%   10.79k ± 0%       ~ (p=0.182 n=6)
Query/nh_100,_instant_query/Mimir-8                                                   19.16k ± 0%   19.16k ± 0%       ~ (p=0.455 n=6)
Query/nh_100,_instant_query/Prometheus-8                                              19.20k ± 0%   19.20k ± 0%       ~ (p=0.242 n=6)
Query/nh_100,_range_query_with_100_steps/Mimir-8                                      120.1k ± 0%   120.1k ± 0%       ~ (p=0.818 n=6)
Query/nh_100,_range_query_with_100_steps/Prometheus-8                                 120.1k ± 0%   120.1k ± 0%       ~ (p=0.998 n=6)
Query/nh_100,_range_query_with_1000_steps/Mimir-8                                     1.027M ± 0%   1.027M ± 0%       ~ (p=0.619 n=6)
Query/nh_100,_range_query_with_1000_steps/Prometheus-8                                1.027M ± 0%   1.027M ± 0%       ~ (p=0.671 n=6)
Query/nh_2000,_instant_query/Mimir-8                                                  375.0k ± 0%   375.0k ± 0%       ~ (p=0.065 n=6)
Query/nh_2000,_instant_query/Prometheus-8                                             375.6k ± 0%   375.5k ± 0%       ~ (p=0.065 n=6)
Query/nh_2000,_range_query_with_100_steps/Mimir-8                                     2.393M ± 0%   2.393M ± 0%       ~ (p=0.937 n=6)
Query/nh_2000,_range_query_with_100_steps/Prometheus-8                                2.393M ± 0%   2.393M ± 0%       ~ (p=0.310 n=6)
Query/nh_2000,_range_query_with_1000_steps/Mimir-8                                    20.54M ± 0%   20.54M ± 0%       ~ (p=0.589 n=6)
Query/nh_2000,_range_query_with_1000_steps/Prometheus-8                               20.54M ± 0%   20.54M ± 0%  +0.00% (p=0.041 n=6)
Query/a_1[1m],_instant_query/Mimir-8                                                   336.0 ± 0%    339.0 ± 0%  +0.89% (p=0.002 n=6)
Query/a_1[1m],_instant_query/Prometheus-8                                              383.0 ± 0%    383.0 ± 0%       ~ (p=1.000 n=6) ¹
Query/a_100[1m],_instant_query/Mimir-8                                                1.845k ± 0%   1.848k ± 0%  +0.16% (p=0.002 n=6)
Query/a_100[1m],_instant_query/Prometheus-8                                           1.906k ± 0%   1.906k ± 0%       ~ (p=1.000 n=6) ¹
Query/a_2000[1m],_instant_query/Mimir-8                                               30.73k ± 0%   30.73k ± 0%       ~ (p=0.052 n=6)
Query/a_2000[1m],_instant_query/Prometheus-8                                          30.80k ± 0%   30.80k ± 0%  -0.01% (p=0.050 n=6)
Query/rate(a_1[1m]),_instant_query/Mimir-8                                             345.0 ± 0%    348.0 ± 0%  +0.87% (p=0.002 n=6)
Query/rate(a_1[1m]),_instant_query/Prometheus-8                                        408.0 ± 0%    408.0 ± 0%       ~ (p=1.000 n=6) ¹
Query/rate(a_1[1m]),_range_query_with_100_steps/Mimir-8                                350.0 ± 0%    353.0 ± 0%  +0.86% (p=0.002 n=6)
Query/rate(a_1[1m]),_range_query_with_100_steps/Prometheus-8                           418.0 ± 0%    418.0 ± 0%       ~ (p=1.000 n=6) ¹
Query/rate(a_1[1m]),_range_query_with_1000_steps/Mimir-8                               383.0 ± 0%    386.0 ± 0%  +0.78% (p=0.002 n=6)
Query/rate(a_1[1m]),_range_query_with_1000_steps/Prometheus-8                          451.0 ± 0%    451.0 ± 0%       ~ (p=1.000 n=6) ¹
Query/rate(a_100[1m]),_instant_query/Mimir-8                                          1.854k ± 0%   1.857k ± 0%  +0.16% (p=0.002 n=6)
Query/rate(a_100[1m]),_instant_query/Prometheus-8                                     2.328k ± 0%   2.328k ± 0%       ~ (p=1.000 n=6) ¹
Query/rate(a_100[1m]),_range_query_with_100_steps/Mimir-8                             2.261k ± 0%   2.264k ± 0%  +0.15% (p=0.002 n=6)
Query/rate(a_100[1m]),_range_query_with_100_steps/Prometheus-8                        2.739k ± 0%   2.739k ± 0%       ~ (p=1.000 n=6)
Query/rate(a_100[1m]),_range_query_with_1000_steps/Mimir-8                            5.574k ± 0%   5.579k ± 0%  +0.09% (p=0.004 n=6)
Query/rate(a_100[1m]),_range_query_with_1000_steps/Prometheus-8                       6.053k ± 0%   6.055k ± 0%       ~ (p=0.684 n=6)
Query/rate(a_2000[1m]),_instant_query/Mimir-8                                         30.72k ± 0%   30.72k ± 0%  +0.01% (p=0.022 n=6)
Query/rate(a_2000[1m]),_instant_query/Prometheus-8                                    38.82k ± 0%   38.82k ± 0%       ~ (p=0.136 n=6)
Query/rate(a_2000[1m]),_range_query_with_100_steps/Mimir-8                            38.78k ± 0%   38.79k ± 0%  +0.03% (p=0.026 n=6)
Query/rate(a_2000[1m]),_range_query_with_100_steps/Prometheus-8                       46.87k ± 0%   46.87k ± 0%       ~ (p=0.506 n=6)
Query/rate(a_2000[1m]),_range_query_with_1000_steps/Mimir-8                           104.9k ± 0%   104.9k ± 0%       ~ (p=0.240 n=6)
Query/rate(a_2000[1m]),_range_query_with_1000_steps/Prometheus-8                      113.0k ± 0%   113.0k ± 0%       ~ (p=0.974 n=6)
Query/rate(a_1[1m]),_range_query_with_10000_steps/Mimir-8                              692.0 ± 0%    695.0 ± 0%  +0.43% (p=0.002 n=6)
Query/rate(a_1[1m]),_range_query_with_10000_steps/Prometheus-8                         760.0 ± 0%    760.0 ± 0%       ~ (p=1.000 n=6)
Query/rate(a_100[1m]),_range_query_with_10000_steps/Mimir-8                           35.94k ± 0%   35.96k ± 0%  +0.06% (p=0.002 n=6)
Query/rate(a_100[1m]),_range_query_with_10000_steps/Prometheus-8                      36.41k ± 0%   36.42k ± 0%       ~ (p=0.461 n=6)
Query/rate(a_2000[1m]),_range_query_with_10000_steps/Mimir-8                          717.9k ± 0%   718.0k ± 0%       ~ (p=0.851 n=6)
Query/rate(a_2000[1m]),_range_query_with_10000_steps/Prometheus-8                     732.2k ± 0%   732.2k ± 0%       ~ (p=0.818 n=6)
Query/rate(a_1[1d]),_instant_query/Mimir-8                                             645.5 ± 0%    648.0 ± 0%  +0.39% (p=0.002 n=6)
Query/rate(a_1[1d]),_instant_query/Prometheus-8                                        722.0 ± 0%    721.5 ± 0%       ~ (p=0.892 n=6)
Query/rate(a_1[1d]),_range_query_with_100_steps/Mimir-8                                648.0 ± 0%    652.0 ± 0%  +0.62% (p=0.002 n=6)
Query/rate(a_1[1d]),_range_query_with_100_steps/Prometheus-8                           730.0 ± 0%    730.5 ± 0%       ~ (p=1.000 n=6)
Query/rate(a_1[1d]),_range_query_with_1000_steps/Mimir-8                               678.0 ± 0%    681.0 ± 1%  +0.44% (p=0.002 n=6)
Query/rate(a_1[1d]),_range_query_with_1000_steps/Prometheus-8                          760.0 ± 0%    759.5 ± 0%       ~ (p=0.961 n=6)
Query/rate(a_100[1d]),_instant_query/Mimir-8                                          31.77k ± 0%   31.77k ± 0%       ~ (p=1.000 n=6)
Query/rate(a_100[1d]),_instant_query/Prometheus-8                                     32.27k ± 0%   32.26k ± 0%       ~ (p=0.937 n=6)
Query/rate(a_100[1d]),_range_query_with_100_steps/Mimir-8                             32.09k ± 0%   32.10k ± 0%       ~ (p=0.790 n=6)
Query/rate(a_100[1d]),_range_query_with_100_steps/Prometheus-8                        32.56k ± 0%   32.55k ± 0%       ~ (p=1.000 n=6)
Query/rate(a_100[1d]),_range_query_with_1000_steps/Mimir-8                            35.16k ± 0%   35.18k ± 0%       ~ (p=0.260 n=6)
Query/rate(a_100[1d]),_range_query_with_1000_steps/Prometheus-8                       37.42k ± 0%   37.44k ± 0%       ~ (p=0.851 n=6)
Query/rate(a_2000[1d]),_instant_query/Mimir-8                                         629.7k ± 0%   629.7k ± 0%       ~ (p=0.900 n=6)
Query/rate(a_2000[1d]),_instant_query/Prometheus-8                                    642.0k ± 0%   641.8k ± 0%  -0.02% (p=0.009 n=6)
Query/rate(a_2000[1d]),_range_query_with_100_steps/Mimir-8                            639.8k ± 0%   639.8k ± 0%       ~ (p=0.699 n=6)
Query/rate(a_2000[1d]),_range_query_with_100_steps/Prometheus-8                       647.9k ± 0%   647.9k ± 0%       ~ (p=0.937 n=6)
Query/rate(a_2000[1d]),_range_query_with_1000_steps/Mimir-8                           700.2k ± 0%   700.2k ± 0%       ~ (p=0.485 n=6)
Query/rate(a_2000[1d]),_range_query_with_1000_steps/Prometheus-8                      708.2k ± 0%   708.2k ± 0%       ~ (p=0.699 n=6)
Query/a_1_-_b_1,_instant_query/Mimir-8                                                 666.0 ± 0%    666.0 ± 0%       ~ (p=1.000 n=6)
Query/a_1_-_b_1,_instant_query/Prometheus-8                                            730.0 ± 0%    730.5 ± 0%       ~ (p=0.545 n=6)
Query/a_1_-_b_1,_range_query_with_100_steps/Mimir-8                                    675.0 ± 0%    675.0 ± 0%       ~ (p=1.000 n=6)
Query/a_1_-_b_1,_range_query_with_100_steps/Prometheus-8                               943.0 ± 0%    942.5 ± 0%       ~ (p=0.545 n=6)
Query/a_1_-_b_1,_range_query_with_1000_steps/Mimir-8                                   739.0 ± 0%    739.0 ± 0%       ~ (p=1.000 n=6) ¹
Query/a_1_-_b_1,_range_query_with_1000_steps/Prometheus-8                             2.807k ± 0%   2.807k ± 0%       ~ (p=1.000 n=6) ¹
Query/a_100_-_b_100,_instant_query/Mimir-8                                            5.287k ± 0%   5.287k ± 0%       ~ (p=1.000 n=6)
Query/a_100_-_b_100,_instant_query/Prometheus-8                                       5.193k ± 0%   5.192k ± 0%       ~ (p=0.123 n=6)
Query/a_100_-_b_100,_range_query_with_100_steps/Mimir-8                               6.101k ± 0%   6.101k ± 0%       ~ (p=0.407 n=6)
Query/a_100_-_b_100,_range_query_with_100_steps/Prometheus-8                          6.214k ± 0%   6.215k ± 0%       ~ (p=0.708 n=6)
Query/a_100_-_b_100,_range_query_with_1000_steps/Mimir-8                              12.52k ± 0%   12.52k ± 0%       ~ (p=0.578 n=6)
Query/a_100_-_b_100,_range_query_with_1000_steps/Prometheus-8                         16.31k ± 0%   16.31k ± 0%       ~ (p=0.216 n=6)
Query/a_2000_-_b_2000,_instant_query/Mimir-8                                          93.55k ± 0%   93.58k ± 0%       ~ (p=0.121 n=6)
Query/a_2000_-_b_2000,_instant_query/Prometheus-8                                     89.71k ± 0%   89.71k ± 0%       ~ (p=0.623 n=6)
Query/a_2000_-_b_2000,_range_query_with_100_steps/Mimir-8                             109.6k ± 0%   109.6k ± 0%       ~ (p=0.087 n=6)
Query/a_2000_-_b_2000,_range_query_with_100_steps/Prometheus-8                        102.0k ± 0%   102.0k ± 1%  +0.04% (p=0.017 n=6)
Query/a_2000_-_b_2000,_range_query_with_1000_steps/Mimir-8                            237.8k ± 0%   237.8k ± 0%       ~ (p=0.699 n=6)
Query/a_2000_-_b_2000,_range_query_with_1000_steps/Prometheus-8                       245.9k ± 0%   245.8k ± 0%       ~ (p=0.665 n=6)
Query/a_1_-_b_1,_range_query_with_10000_steps/Mimir-8                                 1.351k ± 0%   1.351k ± 0%       ~ (p=0.364 n=6)
Query/a_1_-_b_1,_range_query_with_10000_steps/Prometheus-8                            21.42k ± 0%   21.42k ± 0%       ~ (p=1.000 n=6)
Query/a_100_-_b_100,_range_query_with_10000_steps/Mimir-8                             72.63k ± 0%   72.64k ± 0%  +0.02% (p=0.009 n=6)
Query/a_100_-_b_100,_range_query_with_10000_steps/Prometheus-8                        113.2k ± 0%   113.2k ± 0%       ~ (p=0.180 n=6)
Query/a_2000_-_b_2000,_range_query_with_10000_steps/Mimir-8                           1.446M ± 0%   1.446M ± 0%       ~ (p=0.937 n=6)
Query/a_2000_-_b_2000,_range_query_with_10000_steps/Prometheus-8                      1.484M ± 0%   1.484M ± 0%       ~ (p=0.667 n=6)
Query/a_100{l=~"[13579]."}_-_b_100,_instant_query/Mimir-8                             4.032k ± 0%   4.032k ± 0%       ~ (p=0.545 n=6)
Query/a_100{l=~"[13579]."}_-_b_100,_instant_query/Prometheus-8                        4.034k ± 0%   4.034k ± 0%       ~ (p=1.000 n=6) ¹
Query/a_100{l=~"[13579]."}_-_b_100,_range_query_with_100_steps/Mimir-8                4.645k ± 0%   4.646k ± 0%       ~ (p=0.413 n=6)
Query/a_100{l=~"[13579]."}_-_b_100,_range_query_with_100_steps/Prometheus-8           4.877k ± 0%   4.876k ± 0%       ~ (p=0.716 n=6)
Query/a_100{l=~"[13579]."}_-_b_100,_range_query_with_1000_steps/Mimir-8               9.459k ± 0%   9.464k ± 0%       ~ (p=0.097 n=6)
Query/a_100{l=~"[13579]."}_-_b_100,_range_query_with_1000_steps/Prometheus-8          13.10k ± 0%   13.10k ± 0%       ~ (p=0.223 n=6)
Query/a_2000{l=~"1..."}_-_b_2000,_instant_query/Mimir-8                               58.54k ± 0%   58.54k ± 0%       ~ (p=0.853 n=6)
Query/a_2000{l=~"1..."}_-_b_2000,_instant_query/Prometheus-8                          66.49k ± 0%   66.50k ± 0%       ~ (p=0.225 n=6)
Query/a_2000{l=~"1..."}_-_b_2000,_range_query_with_100_steps/Mimir-8                  67.92k ± 0%   67.94k ± 0%       ~ (p=0.258 n=6)
Query/a_2000{l=~"1..."}_-_b_2000,_range_query_with_100_steps/Prometheus-8             75.75k ± 0%   75.76k ± 0%       ~ (p=0.426 n=6)
Query/a_2000{l=~"1..."}_-_b_2000,_range_query_with_1000_steps/Mimir-8                 137.2k ± 0%   137.2k ± 0%       ~ (p=0.485 n=6)
Query/a_2000{l=~"1..."}_-_b_2000,_range_query_with_1000_steps/Prometheus-8            173.8k ± 0%   173.8k ± 0%       ~ (p=0.394 n=6)
Query/a_2000{l="1234"}_-_b_2000,_instant_query/Mimir-8                                23.22k ± 1%   23.16k ± 1%       ~ (p=0.394 n=6)
Query/a_2000{l="1234"}_-_b_2000,_instant_query/Prometheus-8                           43.23k ± 0%   43.23k ± 0%       ~ (p=0.100 n=6)
Query/a_2000{l="1234"}_-_b_2000,_range_query_with_100_steps/Mimir-8                   24.95k ± 3%   25.38k ± 2%       ~ (p=0.589 n=6)
Query/a_2000{l="1234"}_-_b_2000,_range_query_with_100_steps/Prometheus-8              49.50k ± 0%   49.49k ± 0%       ~ (p=1.000 n=6)
Query/a_2000{l="1234"}_-_b_2000,_range_query_with_1000_steps/Mimir-8                  34.03k ± 2%   33.94k ± 1%       ~ (p=0.240 n=6)
Query/a_2000{l="1234"}_-_b_2000,_range_query_with_1000_steps/Prometheus-8             115.4k ± 0%   115.4k ± 0%       ~ (p=0.699 n=6)
Query/a_2000_-_b_2000{l="1234"},_instant_query/Mimir-8                                23.63k ± 0%   23.56k ± 0%       ~ (p=0.065 n=6)
Query/a_2000_-_b_2000{l="1234"},_instant_query/Prometheus-8                           43.16k ± 0%   43.17k ± 0%       ~ (p=0.333 n=6)
Query/a_2000_-_b_2000{l="1234"},_range_query_with_100_steps/Mimir-8                   25.43k ± 3%   25.00k ± 4%       ~ (p=0.310 n=6)
Query/a_2000_-_b_2000{l="1234"},_range_query_with_100_steps/Prometheus-8              49.43k ± 0%   49.43k ± 0%       ~ (p=0.699 n=6)
Query/a_2000_-_b_2000{l="1234"},_range_query_with_1000_steps/Mimir-8                  34.06k ± 2%   34.08k ± 1%       ~ (p=0.818 n=6)
Query/a_2000_-_b_2000{l="1234"},_range_query_with_1000_steps/Prometheus-8             115.4k ± 0%   115.4k ± 0%       ~ (p=0.558 n=6)
Query/sum(a_1),_instant_query/Mimir-8                                                  347.0 ± 0%    347.0 ± 0%       ~ (p=1.000 n=6) ¹
Query/sum(a_1),_instant_query/Prometheus-8                                             391.0 ± 0%    391.0 ± 0%       ~ (p=1.000 n=6) ¹
Query/sum(a_1),_range_query_with_100_steps/Mimir-8                                     352.0 ± 0%    352.0 ± 0%       ~ (p=1.000 n=6) ¹
Query/sum(a_1),_range_query_with_100_steps/Prometheus-8                                401.0 ± 0%    401.0 ± 0%       ~ (p=1.000 n=6) ¹
Query/sum(a_1),_range_query_with_1000_steps/Mimir-8                                    384.0 ± 0%    384.0 ± 0%       ~ (p=1.000 n=6) ¹
Query/sum(a_1),_range_query_with_1000_steps/Prometheus-8                               433.0 ± 0%    433.0 ± 0%       ~ (p=1.000 n=6) ¹
Query/sum(a_100),_instant_query/Mimir-8                                               2.257k ± 0%   2.257k ± 0%       ~ (p=1.000 n=6) ¹
Query/sum(a_100),_instant_query/Prometheus-8                                          2.308k ± 0%   2.309k ± 0%       ~ (p=0.545 n=6)
Query/sum(a_100),_range_query_with_100_steps/Mimir-8                                  2.664k ± 0%   2.664k ± 0%       ~ (p=1.000 n=6)
Query/sum(a_100),_range_query_with_100_steps/Prometheus-8                             2.720k ± 0%   2.720k ± 0%       ~ (p=0.455 n=6)
Query/sum(a_100),_range_query_with_1000_steps/Mimir-8                                 5.871k ± 0%   5.871k ± 0%       ~ (p=0.768 n=6)
Query/sum(a_100),_range_query_with_1000_steps/Prometheus-8                            5.925k ± 0%   5.927k ± 0%       ~ (p=0.177 n=6)
Query/sum(a_2000),_instant_query/Mimir-8                                              38.76k ± 0%   38.77k ± 0%  +0.02% (p=0.011 n=6)
Query/sum(a_2000),_instant_query/Prometheus-8                                         38.83k ± 0%   38.84k ± 0%       ~ (p=0.074 n=6)
Query/sum(a_2000),_range_query_with_100_steps/Mimir-8                                 46.80k ± 0%   46.81k ± 0%       ~ (p=0.082 n=6)
Query/sum(a_2000),_range_query_with_100_steps/Prometheus-8                            46.87k ± 0%   46.89k ± 0%       ~ (p=0.162 n=6)
Query/sum(a_2000),_range_query_with_1000_steps/Mimir-8                                110.9k ± 0%   110.9k ± 0%       ~ (p=0.818 n=6)
Query/sum(a_2000),_range_query_with_1000_steps/Prometheus-8                           111.0k ± 0%   111.0k ± 0%       ~ (p=0.818 n=6)
Query/sum_by_(l)(h_1),_instant_query/Mimir-8                                           454.0 ± 0%    454.0 ± 0%       ~ (p=1.000 n=6) ¹
Query/sum_by_(l)(h_1),_instant_query/Prometheus-8                                      504.0 ± 0%    504.0 ± 0%       ~ (p=1.000 n=6) ¹
Query/sum_by_(l)(h_1),_range_query_with_100_steps/Mimir-8                              479.0 ± 0%    479.0 ± 0%       ~ (p=1.000 n=6)
Query/sum_by_(l)(h_1),_range_query_with_100_steps/Prometheus-8                         534.0 ± 0%    534.0 ± 0%       ~ (p=1.000 n=6) ¹
Query/sum_by_(l)(h_1),_range_query_with_1000_steps/Mimir-8                             672.0 ± 0%    672.0 ± 0%       ~ (p=1.000 n=6) ¹
Query/sum_by_(l)(h_1),_range_query_with_1000_steps/Prometheus-8                        727.0 ± 0%    727.0 ± 0%       ~ (p=1.000 n=6)
Query/sum_by_(l)(h_100),_instant_query/Mimir-8                                        12.59k ± 0%   12.59k ± 0%       ~ (p=0.182 n=6)
Query/sum_by_(l)(h_100),_instant_query/Prometheus-8                                   12.66k ± 0%   12.66k ± 0%       ~ (p=0.448 n=6)
Query/sum_by_(l)(h_100),_range_query_with_100_steps/Mimir-8                           15.01k ± 0%   15.01k ± 0%       ~ (p=0.684 n=6)
Query/sum_by_(l)(h_100),_range_query_with_100_steps/Prometheus-8                      15.07k ± 0%   15.07k ± 0%       ~ (p=0.844 n=6)
Query/sum_by_(l)(h_100),_range_query_with_1000_steps/Mimir-8                          34.28k ± 0%   34.28k ± 0%       ~ (p=1.000 n=6)
Query/sum_by_(l)(h_100),_range_query_with_1000_steps/Prometheus-8                     34.33k ± 0%   34.33k ± 0%       ~ (p=0.574 n=6)
Query/sum_by_(l)(h_2000),_instant_query/Mimir-8                                       245.5k ± 0%   246.0k ± 0%       ~ (p=0.065 n=6)
Query/sum_by_(l)(h_2000),_instant_query/Prometheus-8                                  258.9k ± 2%   256.4k ± 2%  -0.95% (p=0.041 n=6)
Query/sum_by_(l)(h_2000),_range_query_with_100_steps/Mimir-8                          295.9k ± 0%   295.3k ± 1%       ~ (p=0.699 n=6)
Query/sum_by_(l)(h_2000),_range_query_with_100_steps/Prometheus-8                     293.2k ± 0%   293.3k ± 0%       ~ (p=0.485 n=6)
Query/sum_by_(l)(h_2000),_range_query_with_1000_steps/Mimir-8                         689.3k ± 0%   689.3k ± 0%       ~ (p=0.506 n=6)
Query/sum_by_(l)(h_2000),_range_query_with_1000_steps/Prometheus-8                    707.6k ± 0%   707.6k ± 0%       ~ (p=0.394 n=6)
Query/sum_by_(le)(h_1),_instant_query/Mimir-8                                          460.0 ± 0%    460.0 ± 0%       ~ (p=1.000 n=6) ¹
Query/sum_by_(le)(h_1),_instant_query/Prometheus-8                                     512.0 ± 0%    512.0 ± 0%       ~ (p=1.000 n=6) ¹
Query/sum_by_(le)(h_1),_range_query_with_100_steps/Mimir-8                             484.5 ± 0%    484.0 ± 0%       ~ (p=0.545 n=6)
Query/sum_by_(le)(h_1),_range_query_with_100_steps/Prometheus-8                        542.0 ± 0%    542.0 ± 0%       ~ (p=1.000 n=6)
Query/sum_by_(le)(h_1),_range_query_with_1000_steps/Mimir-8                            677.0 ± 0%    677.0 ± 0%       ~ (p=1.000 n=6) ¹
Query/sum_by_(le)(h_1),_range_query_with_1000_steps/Prometheus-8                       735.0 ± 0%    735.0 ± 0%       ~ (p=1.000 n=6) ¹
Query/sum_by_(le)(h_100),_instant_query/Mimir-8                                       12.48k ± 0%   12.48k ± 0%       ~ (p=1.000 n=6)
Query/sum_by_(le)(h_100),_instant_query/Prometheus-8                                  12.55k ± 0%   12.55k ± 0%       ~ (p=0.316 n=6)
Query/sum_by_(le)(h_100),_range_query_with_100_steps/Mimir-8                          14.90k ± 0%   14.90k ± 0%       ~ (p=0.491 n=6)
Query/sum_by_(le)(h_100),_range_query_with_100_steps/Prometheus-8                     14.96k ± 0%   14.96k ± 0%       ~ (p=0.727 n=6)
Query/sum_by_(le)(h_100),_range_query_with_1000_steps/Mimir-8                         34.16k ± 0%   34.16k ± 0%       ~ (p=0.615 n=6)
Query/sum_by_(le)(h_100),_range_query_with_1000_steps/Prometheus-8                    34.20k ± 0%   34.21k ± 0%       ~ (p=1.000 n=6)
Query/sum_by_(le)(h_2000),_instant_query/Mimir-8                                      242.9k ± 0%   243.0k ± 0%       ~ (p=0.372 n=6)
Query/sum_by_(le)(h_2000),_instant_query/Prometheus-8                                 255.5k ± 1%   256.7k ± 1%       ~ (p=0.485 n=6)
Query/sum_by_(le)(h_2000),_range_query_with_100_steps/Mimir-8                         291.2k ± 0%   291.1k ± 0%       ~ (p=0.394 n=6)
Query/sum_by_(le)(h_2000),_range_query_with_100_steps/Prometheus-8                    291.2k ± 0%   291.2k ± 0%       ~ (p=0.818 n=6)
Query/sum_by_(le)(h_2000),_range_query_with_1000_steps/Mimir-8                        677.6k ± 0%   677.6k ± 0%       ~ (p=0.485 n=6)
Query/sum_by_(le)(h_2000),_range_query_with_1000_steps/Prometheus-8                   701.5k ± 0%   701.5k ± 0%       ~ (p=0.394 n=6)
Query/rate(a_1[1m])_+_rate(b_1[1m]),_instant_query/Mimir-8                             678.0 ± 0%    684.0 ± 0%  +0.88% (p=0.002 n=6)
Query/rate(a_1[1m])_+_rate(b_1[1m]),_instant_query/Prometheus-8                        805.0 ± 0%    805.0 ± 0%       ~ (p=1.000 n=6) ¹
Query/rate(a_1[1m])_+_rate(b_1[1m]),_range_query_with_100_steps/Mimir-8                686.0 ± 0%    692.0 ± 0%  +0.87% (p=0.002 n=6)
Query/rate(a_1[1m])_+_rate(b_1[1m]),_range_query_with_100_steps/Prometheus-8          1.017k ± 0%   1.017k ± 0%       ~ (p=1.000 n=6) ¹
Query/rate(a_1[1m])_+_rate(b_1[1m]),_range_query_with_1000_steps/Mimir-8               752.0 ± 0%    758.0 ± 0%  +0.80% (p=0.002 n=6)
Query/rate(a_1[1m])_+_rate(b_1[1m]),_range_query_with_1000_steps/Prometheus-8         2.884k ± 0%   2.884k ± 0%       ~ (p=1.000 n=6)
Query/rate(a_100[1m])_+_rate(b_100[1m]),_instant_query/Mimir-8                        4.499k ± 0%   4.505k ± 0%  +0.13% (p=0.002 n=6)
Query/rate(a_100[1m])_+_rate(b_100[1m]),_instant_query/Prometheus-8                   5.275k ± 0%   5.276k ± 0%       ~ (p=1.000 n=6)
Query/rate(a_100[1m])_+_rate(b_100[1m]),_range_query_with_100_steps/Mimir-8           5.312k ± 0%   5.318k ± 0%  +0.11% (p=0.002 n=6)
Query/rate(a_100[1m])_+_rate(b_100[1m]),_range_query_with_100_steps/Prometheus-8      6.297k ± 0%   6.297k ± 0%       ~ (p=0.623 n=6)
Query/rate(a_100[1m])_+_rate(b_100[1m]),_range_query_with_1000_steps/Mimir-8          11.95k ± 0%   11.95k ± 0%  +0.04% (p=0.045 n=6)
Query/rate(a_100[1m])_+_rate(b_100[1m]),_range_query_with_1000_steps/Prometheus-8     16.61k ± 0%   16.60k ± 0%       ~ (p=0.234 n=6)
Query/rate(a_2000[1m])_+_rate(b_2000[1m]),_instant_query/Mimir-8                      77.51k ± 0%   77.51k ± 0%       ~ (p=0.190 n=6)
Query/rate(a_2000[1m])_+_rate(b_2000[1m]),_instant_query/Prometheus-8                 89.73k ± 0%   89.74k ± 0%       ~ (p=0.199 n=6)
Query/rate(a_2000[1m])_+_rate(b_2000[1m]),_range_query_with_100_steps/Mimir-8         93.61k ± 0%   93.63k ± 0%       ~ (p=0.065 n=6)
Query/rate(a_2000[1m])_+_rate(b_2000[1m]),_range_query_with_100_steps/Prometheus-8    102.0k ± 0%   102.0k ± 0%       ~ (p=0.394 n=6)
Query/rate(a_2000[1m])_+_rate(b_2000[1m]),_range_query_with_1000_steps/Mimir-8        226.0k ± 0%   226.0k ± 0%       ~ (p=0.784 n=6)
Query/rate(a_2000[1m])_+_rate(b_2000[1m]),_range_query_with_1000_steps/Prometheus-8   250.0k ± 0%   250.0k ± 0%       ~ (p=1.000 n=6)
Query/sum(a_1_+_b_1),_instant_query/Mimir-8                                            677.0 ± 0%    677.0 ± 0%       ~ (p=1.000 n=6) ¹
Query/sum(a_1_+_b_1),_instant_query/Prometheus-8                                       756.5 ± 0%    757.0 ± 0%       ~ (p=1.000 n=6)
Query/sum(a_1_+_b_1),_range_query_with_100_steps/Mimir-8                               686.0 ± 0%    686.0 ± 0%       ~ (p=1.000 n=6) ¹
Query/sum(a_1_+_b_1),_range_query_with_100_steps/Prometheus-8                          969.0 ± 0%    969.0 ± 0%       ~ (p=1.000 n=6)
Query/sum(a_1_+_b_1),_range_query_with_1000_steps/Mimir-8                              750.0 ± 0%    750.0 ± 0%       ~ (p=1.000 n=6)
Query/sum(a_1_+_b_1),_range_query_with_1000_steps/Prometheus-8                        2.833k ± 0%   2.833k ± 0%       ~ (p=1.000 n=6)
Query/sum(a_100_+_b_100),_instant_query/Mimir-8                                       5.300k ± 0%   5.300k ± 0%       ~ (p=0.455 n=6)
Query/sum(a_100_+_b_100),_instant_query/Prometheus-8                                  5.218k ± 0%   5.219k ± 0%       ~ (p=0.405 n=6)
Query/sum(a_100_+_b_100),_range_query_with_100_steps/Mimir-8                          6.112k ± 0%   6.111k ± 0%       ~ (p=0.675 n=6)
Query/sum(a_100_+_b_100),_range_query_with_100_steps/Prometheus-8                     6.238k ± 0%   6.240k ± 0%       ~ (p=0.413 n=6)
Query/sum(a_100_+_b_100),_range_query_with_1000_steps/Mimir-8                         12.53k ± 0%   12.53k ± 0%       ~ (p=0.554 n=6)
Query/sum(a_100_+_b_100),_range_query_with_1000_steps/Prometheus-8                    16.34k ± 0%   16.35k ± 0%       ~ (p=0.143 n=6)
Query/sum(a_2000_+_b_2000),_instant_query/Mimir-8                                     93.58k ± 0%   93.57k ± 0%       ~ (p=0.968 n=6)
Query/sum(a_2000_+_b_2000),_instant_query/Prometheus-8                                89.73k ± 0%   89.73k ± 0%       ~ (p=0.738 n=6)
Query/sum(a_2000_+_b_2000),_range_query_with_100_steps/Mimir-8                        109.7k ± 0%   109.6k ± 0%       ~ (p=0.855 n=6)
Query/sum(a_2000_+_b_2000),_range_query_with_100_steps/Prometheus-8                   102.0k ± 2%   102.0k ± 0%       ~ (p=0.515 n=6)
Query/sum(a_2000_+_b_2000),_range_query_with_1000_steps/Mimir-8                       237.9k ± 0%   237.9k ± 0%       ~ (p=0.329 n=6)
Query/sum(a_2000_+_b_2000),_range_query_with_1000_steps/Prometheus-8                  245.9k ± 0%   245.9k ± 0%       ~ (p=0.485 n=6)
Query/sum_by_(le)(rate(h_1[1m])),_instant_query/Mimir-8                                443.0 ± 0%    446.0 ± 0%  +0.68% (p=0.002 n=6)
Query/sum_by_(le)(rate(h_1[1m])),_instant_query/Prometheus-8                           556.0 ± 0%    556.0 ± 0%       ~ (p=1.000 n=6) ¹
Query/sum_by_(le)(rate(h_1[1m])),_range_query_with_100_steps/Mimir-8                   467.0 ± 0%    470.0 ± 0%  +0.64% (p=0.002 n=6)
Query/sum_by_(le)(rate(h_1[1m])),_range_query_with_100_steps/Prometheus-8              578.0 ± 0%    578.0 ± 0%       ~ (p=0.455 n=6)
Query/sum_by_(le)(rate(h_1[1m])),_range_query_with_1000_steps/Mimir-8                  666.0 ± 0%    669.0 ± 0%  +0.45% (p=0.002 n=6)
Query/sum_by_(le)(rate(h_1[1m])),_range_query_with_1000_steps/Prometheus-8             777.0 ± 0%    777.0 ± 0%       ~ (p=1.000 n=6) ¹
Query/sum_by_(le)(rate(h_100[1m])),_instant_query/Mimir-8                             10.08k ± 0%   10.09k ± 0%  +0.03% (p=0.002 n=6)
Query/sum_by_(le)(rate(h_100[1m])),_instant_query/Prometheus-8                        12.59k ± 0%   12.59k ± 0%       ~ (p=0.864 n=6)
Query/sum_by_(le)(rate(h_100[1m])),_range_query_with_100_steps/Mimir-8                12.50k ± 0%   12.51k ± 0%       ~ (p=0.071 n=6)
Query/sum_by_(le)(rate(h_100[1m])),_range_query_with_100_steps/Prometheus-8           15.01k ± 0%   15.01k ± 0%       ~ (p=0.647 n=6)
Query/sum_by_(le)(rate(h_100[1m])),_range_query_with_1000_steps/Mimir-8               32.37k ± 0%   32.38k ± 0%       ~ (p=0.087 n=6)
Query/sum_by_(le)(rate(h_100[1m])),_range_query_with_1000_steps/Prometheus-8          34.88k ± 0%   34.88k ± 0%       ~ (p=0.431 n=6)
Query/sum_by_(le)(rate(h_2000[1m])),_instant_query/Mimir-8                            194.7k ± 0%   194.7k ± 0%       ~ (p=0.818 n=6)
Query/sum_by_(le)(rate(h_2000[1m])),_instant_query/Prometheus-8                       252.5k ± 1%   250.4k ± 0%  -0.83% (p=0.041 n=6)
Query/sum_by_(le)(rate(h_2000[1m])),_range_query_with_100_steps/Mimir-8               243.0k ± 0%   243.1k ± 0%       ~ (p=0.937 n=6)
Query/sum_by_(le)(rate(h_2000[1m])),_range_query_with_100_steps/Prometheus-8          291.2k ± 0%   291.2k ± 0%       ~ (p=0.937 n=6)
Query/sum_by_(le)(rate(h_2000[1m])),_range_query_with_1000_steps/Mimir-8              641.6k ± 0%   641.6k ± 0%       ~ (p=0.240 n=6)
Query/sum_by_(le)(rate(h_2000[1m])),_range_query_with_1000_steps/Prometheus-8         713.7k ± 0%   713.7k ± 0%       ~ (p=0.485 n=6)
geomean                                                                               13.34k        13.35k       +0.04%
¹ all samples are equal

                                                                                    │   main.txt   │  mqe-histogram-range-vectors.txt   │
                                                                                    │      B       │      B        vs base              │
Query/a_1,_instant_query/Mimir-8                                                      831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_1,_instant_query/Prometheus-8                                                 831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_1,_range_query_with_100_steps/Mimir-8                                         831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_1,_range_query_with_100_steps/Prometheus-8                                    831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_1,_range_query_with_1000_steps/Mimir-8                                        831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_1,_range_query_with_1000_steps/Prometheus-8                                   831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_100,_instant_query/Mimir-8                                                    831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_100,_instant_query/Prometheus-8                                               831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_100,_range_query_with_100_steps/Mimir-8                                       831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_100,_range_query_with_100_steps/Prometheus-8                                  831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_100,_range_query_with_1000_steps/Mimir-8                                      831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_100,_range_query_with_1000_steps/Prometheus-8                                 831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_2000,_instant_query/Mimir-8                                                   831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_2000,_instant_query/Prometheus-8                                              831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_2000,_range_query_with_100_steps/Mimir-8                                      831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_2000,_range_query_with_100_steps/Prometheus-8                                 831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_2000,_range_query_with_1000_steps/Mimir-8                                     831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_2000,_range_query_with_1000_steps/Prometheus-8                                831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/nh_1,_instant_query/Mimir-8                                                     831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/nh_1,_instant_query/Prometheus-8                                                831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/nh_1,_range_query_with_100_steps/Mimir-8                                        831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/nh_1,_range_query_with_100_steps/Prometheus-8                                   831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/nh_1,_range_query_with_1000_steps/Mimir-8                                       831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/nh_1,_range_query_with_1000_steps/Prometheus-8                                  831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/nh_100,_instant_query/Mimir-8                                                   831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/nh_100,_instant_query/Prometheus-8                                              831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/nh_100,_range_query_with_100_steps/Mimir-8                                      831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/nh_100,_range_query_with_100_steps/Prometheus-8                                 831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/nh_100,_range_query_with_1000_steps/Mimir-8                                     831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/nh_100,_range_query_with_1000_steps/Prometheus-8                                831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/nh_2000,_instant_query/Mimir-8                                                  831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/nh_2000,_instant_query/Prometheus-8                                             831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/nh_2000,_range_query_with_100_steps/Mimir-8                                     831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/nh_2000,_range_query_with_100_steps/Prometheus-8                                831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/nh_2000,_range_query_with_1000_steps/Mimir-8                                    831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/nh_2000,_range_query_with_1000_steps/Prometheus-8                               831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_1[1m],_instant_query/Mimir-8                                                  831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_1[1m],_instant_query/Prometheus-8                                             831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_100[1m],_instant_query/Mimir-8                                                831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_100[1m],_instant_query/Prometheus-8                                           831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_2000[1m],_instant_query/Mimir-8                                               831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_2000[1m],_instant_query/Prometheus-8                                          831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_1[1m]),_instant_query/Mimir-8                                            831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_1[1m]),_instant_query/Prometheus-8                                       831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_1[1m]),_range_query_with_100_steps/Mimir-8                               831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_1[1m]),_range_query_with_100_steps/Prometheus-8                          831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_1[1m]),_range_query_with_1000_steps/Mimir-8                              831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_1[1m]),_range_query_with_1000_steps/Prometheus-8                         831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_100[1m]),_instant_query/Mimir-8                                          831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_100[1m]),_instant_query/Prometheus-8                                     831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_100[1m]),_range_query_with_100_steps/Mimir-8                             831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_100[1m]),_range_query_with_100_steps/Prometheus-8                        831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_100[1m]),_range_query_with_1000_steps/Mimir-8                            831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_100[1m]),_range_query_with_1000_steps/Prometheus-8                       831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_2000[1m]),_instant_query/Mimir-8                                         831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_2000[1m]),_instant_query/Prometheus-8                                    831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_2000[1m]),_range_query_with_100_steps/Mimir-8                            831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_2000[1m]),_range_query_with_100_steps/Prometheus-8                       831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_2000[1m]),_range_query_with_1000_steps/Mimir-8                           831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_2000[1m]),_range_query_with_1000_steps/Prometheus-8                      831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_1[1m]),_range_query_with_10000_steps/Mimir-8                             831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_1[1m]),_range_query_with_10000_steps/Prometheus-8                        831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_100[1m]),_range_query_with_10000_steps/Mimir-8                           831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_100[1m]),_range_query_with_10000_steps/Prometheus-8                      831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_2000[1m]),_range_query_with_10000_steps/Mimir-8                          831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_2000[1m]),_range_query_with_10000_steps/Prometheus-8                     831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_1[1d]),_instant_query/Mimir-8                                            831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_1[1d]),_instant_query/Prometheus-8                                       831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_1[1d]),_range_query_with_100_steps/Mimir-8                               831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_1[1d]),_range_query_with_100_steps/Prometheus-8                          831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_1[1d]),_range_query_with_1000_steps/Mimir-8                              831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_1[1d]),_range_query_with_1000_steps/Prometheus-8                         831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_100[1d]),_instant_query/Mimir-8                                          831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_100[1d]),_instant_query/Prometheus-8                                     831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_100[1d]),_range_query_with_100_steps/Mimir-8                             831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_100[1d]),_range_query_with_100_steps/Prometheus-8                        831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_100[1d]),_range_query_with_1000_steps/Mimir-8                            831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_100[1d]),_range_query_with_1000_steps/Prometheus-8                       831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_2000[1d]),_instant_query/Mimir-8                                         831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_2000[1d]),_instant_query/Prometheus-8                                    831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_2000[1d]),_range_query_with_100_steps/Mimir-8                            831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_2000[1d]),_range_query_with_100_steps/Prometheus-8                       831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_2000[1d]),_range_query_with_1000_steps/Mimir-8                           831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_2000[1d]),_range_query_with_1000_steps/Prometheus-8                      831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_1_-_b_1,_instant_query/Mimir-8                                                831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_1_-_b_1,_instant_query/Prometheus-8                                           831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_1_-_b_1,_range_query_with_100_steps/Mimir-8                                   831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_1_-_b_1,_range_query_with_100_steps/Prometheus-8                              831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_1_-_b_1,_range_query_with_1000_steps/Mimir-8                                  831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_1_-_b_1,_range_query_with_1000_steps/Prometheus-8                             831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_100_-_b_100,_instant_query/Mimir-8                                            831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_100_-_b_100,_instant_query/Prometheus-8                                       831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_100_-_b_100,_range_query_with_100_steps/Mimir-8                               831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_100_-_b_100,_range_query_with_100_steps/Prometheus-8                          831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_100_-_b_100,_range_query_with_1000_steps/Mimir-8                              831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_100_-_b_100,_range_query_with_1000_steps/Prometheus-8                         831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_2000_-_b_2000,_instant_query/Mimir-8                                          831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_2000_-_b_2000,_instant_query/Prometheus-8                                     831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_2000_-_b_2000,_range_query_with_100_steps/Mimir-8                             831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_2000_-_b_2000,_range_query_with_100_steps/Prometheus-8                        831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_2000_-_b_2000,_range_query_with_1000_steps/Mimir-8                            831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_2000_-_b_2000,_range_query_with_1000_steps/Prometheus-8                       831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_1_-_b_1,_range_query_with_10000_steps/Mimir-8                                 831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_1_-_b_1,_range_query_with_10000_steps/Prometheus-8                            831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_100_-_b_100,_range_query_with_10000_steps/Mimir-8                             831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_100_-_b_100,_range_query_with_10000_steps/Prometheus-8                        831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_2000_-_b_2000,_range_query_with_10000_steps/Mimir-8                           831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_2000_-_b_2000,_range_query_with_10000_steps/Prometheus-8                      1.589Mi ± 2%   1.574Mi ± 2%       ~ (p=0.180 n=6)
Query/a_100{l=~"[13579]."}_-_b_100,_instant_query/Mimir-8                             831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_100{l=~"[13579]."}_-_b_100,_instant_query/Prometheus-8                        831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_100{l=~"[13579]."}_-_b_100,_range_query_with_100_steps/Mimir-8                831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_100{l=~"[13579]."}_-_b_100,_range_query_with_100_steps/Prometheus-8           831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_100{l=~"[13579]."}_-_b_100,_range_query_with_1000_steps/Mimir-8               831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_100{l=~"[13579]."}_-_b_100,_range_query_with_1000_steps/Prometheus-8          831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_2000{l=~"1..."}_-_b_2000,_instant_query/Mimir-8                               831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_2000{l=~"1..."}_-_b_2000,_instant_query/Prometheus-8                          831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_2000{l=~"1..."}_-_b_2000,_range_query_with_100_steps/Mimir-8                  831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_2000{l=~"1..."}_-_b_2000,_range_query_with_100_steps/Prometheus-8             831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_2000{l=~"1..."}_-_b_2000,_range_query_with_1000_steps/Mimir-8                 831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_2000{l=~"1..."}_-_b_2000,_range_query_with_1000_steps/Prometheus-8            831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_2000{l="1234"}_-_b_2000,_instant_query/Mimir-8                                831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_2000{l="1234"}_-_b_2000,_instant_query/Prometheus-8                           831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_2000{l="1234"}_-_b_2000,_range_query_with_100_steps/Mimir-8                   831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_2000{l="1234"}_-_b_2000,_range_query_with_100_steps/Prometheus-8              831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_2000{l="1234"}_-_b_2000,_range_query_with_1000_steps/Mimir-8                  831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_2000{l="1234"}_-_b_2000,_range_query_with_1000_steps/Prometheus-8             831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_2000_-_b_2000{l="1234"},_instant_query/Mimir-8                                831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_2000_-_b_2000{l="1234"},_instant_query/Prometheus-8                           831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_2000_-_b_2000{l="1234"},_range_query_with_100_steps/Mimir-8                   831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_2000_-_b_2000{l="1234"},_range_query_with_100_steps/Prometheus-8              831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_2000_-_b_2000{l="1234"},_range_query_with_1000_steps/Mimir-8                  831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/a_2000_-_b_2000{l="1234"},_range_query_with_1000_steps/Prometheus-8             831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum(a_1),_instant_query/Mimir-8                                                 831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum(a_1),_instant_query/Prometheus-8                                            831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum(a_1),_range_query_with_100_steps/Mimir-8                                    831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum(a_1),_range_query_with_100_steps/Prometheus-8                               831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum(a_1),_range_query_with_1000_steps/Mimir-8                                   831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum(a_1),_range_query_with_1000_steps/Prometheus-8                              831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum(a_100),_instant_query/Mimir-8                                               831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum(a_100),_instant_query/Prometheus-8                                          831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum(a_100),_range_query_with_100_steps/Mimir-8                                  831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum(a_100),_range_query_with_100_steps/Prometheus-8                             831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum(a_100),_range_query_with_1000_steps/Mimir-8                                 831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum(a_100),_range_query_with_1000_steps/Prometheus-8                            831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum(a_2000),_instant_query/Mimir-8                                              831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum(a_2000),_instant_query/Prometheus-8                                         831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum(a_2000),_range_query_with_100_steps/Mimir-8                                 831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum(a_2000),_range_query_with_100_steps/Prometheus-8                            831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum(a_2000),_range_query_with_1000_steps/Mimir-8                                831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum(a_2000),_range_query_with_1000_steps/Prometheus-8                           831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(l)(h_1),_instant_query/Mimir-8                                          831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(l)(h_1),_instant_query/Prometheus-8                                     831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(l)(h_1),_range_query_with_100_steps/Mimir-8                             831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(l)(h_1),_range_query_with_100_steps/Prometheus-8                        831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(l)(h_1),_range_query_with_1000_steps/Mimir-8                            831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(l)(h_1),_range_query_with_1000_steps/Prometheus-8                       831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(l)(h_100),_instant_query/Mimir-8                                        831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(l)(h_100),_instant_query/Prometheus-8                                   831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(l)(h_100),_range_query_with_100_steps/Mimir-8                           831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(l)(h_100),_range_query_with_100_steps/Prometheus-8                      831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(l)(h_100),_range_query_with_1000_steps/Mimir-8                          831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(l)(h_100),_range_query_with_1000_steps/Prometheus-8                     831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(l)(h_2000),_instant_query/Mimir-8                                       831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(l)(h_2000),_instant_query/Prometheus-8                                  831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(l)(h_2000),_range_query_with_100_steps/Mimir-8                          831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(l)(h_2000),_range_query_with_100_steps/Prometheus-8                     831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(l)(h_2000),_range_query_with_1000_steps/Mimir-8                         831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(l)(h_2000),_range_query_with_1000_steps/Prometheus-8                    831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(le)(h_1),_instant_query/Mimir-8                                         831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(le)(h_1),_instant_query/Prometheus-8                                    831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(le)(h_1),_range_query_with_100_steps/Mimir-8                            831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(le)(h_1),_range_query_with_100_steps/Prometheus-8                       831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(le)(h_1),_range_query_with_1000_steps/Mimir-8                           831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(le)(h_1),_range_query_with_1000_steps/Prometheus-8                      831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(le)(h_100),_instant_query/Mimir-8                                       831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(le)(h_100),_instant_query/Prometheus-8                                  831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(le)(h_100),_range_query_with_100_steps/Mimir-8                          831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(le)(h_100),_range_query_with_100_steps/Prometheus-8                     831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(le)(h_100),_range_query_with_1000_steps/Mimir-8                         831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(le)(h_100),_range_query_with_1000_steps/Prometheus-8                    831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(le)(h_2000),_instant_query/Mimir-8                                      831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(le)(h_2000),_instant_query/Prometheus-8                                 831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(le)(h_2000),_range_query_with_100_steps/Mimir-8                         831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(le)(h_2000),_range_query_with_100_steps/Prometheus-8                    831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(le)(h_2000),_range_query_with_1000_steps/Mimir-8                        831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(le)(h_2000),_range_query_with_1000_steps/Prometheus-8                   831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_1[1m])_+_rate(b_1[1m]),_instant_query/Mimir-8                            831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_1[1m])_+_rate(b_1[1m]),_instant_query/Prometheus-8                       831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_1[1m])_+_rate(b_1[1m]),_range_query_with_100_steps/Mimir-8               831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_1[1m])_+_rate(b_1[1m]),_range_query_with_100_steps/Prometheus-8          831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_1[1m])_+_rate(b_1[1m]),_range_query_with_1000_steps/Mimir-8              831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_1[1m])_+_rate(b_1[1m]),_range_query_with_1000_steps/Prometheus-8         831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_100[1m])_+_rate(b_100[1m]),_instant_query/Mimir-8                        831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_100[1m])_+_rate(b_100[1m]),_instant_query/Prometheus-8                   831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_100[1m])_+_rate(b_100[1m]),_range_query_with_100_steps/Mimir-8           831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_100[1m])_+_rate(b_100[1m]),_range_query_with_100_steps/Prometheus-8      831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_100[1m])_+_rate(b_100[1m]),_range_query_with_1000_steps/Mimir-8          831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_100[1m])_+_rate(b_100[1m]),_range_query_with_1000_steps/Prometheus-8     831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_2000[1m])_+_rate(b_2000[1m]),_instant_query/Mimir-8                      831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_2000[1m])_+_rate(b_2000[1m]),_instant_query/Prometheus-8                 831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_2000[1m])_+_rate(b_2000[1m]),_range_query_with_100_steps/Mimir-8         831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_2000[1m])_+_rate(b_2000[1m]),_range_query_with_100_steps/Prometheus-8    831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_2000[1m])_+_rate(b_2000[1m]),_range_query_with_1000_steps/Mimir-8        831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/rate(a_2000[1m])_+_rate(b_2000[1m]),_range_query_with_1000_steps/Prometheus-8   831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum(a_1_+_b_1),_instant_query/Mimir-8                                           831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum(a_1_+_b_1),_instant_query/Prometheus-8                                      831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum(a_1_+_b_1),_range_query_with_100_steps/Mimir-8                              831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum(a_1_+_b_1),_range_query_with_100_steps/Prometheus-8                         831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum(a_1_+_b_1),_range_query_with_1000_steps/Mimir-8                             831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum(a_1_+_b_1),_range_query_with_1000_steps/Prometheus-8                        831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum(a_100_+_b_100),_instant_query/Mimir-8                                       831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum(a_100_+_b_100),_instant_query/Prometheus-8                                  831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum(a_100_+_b_100),_range_query_with_100_steps/Mimir-8                          831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum(a_100_+_b_100),_range_query_with_100_steps/Prometheus-8                     831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum(a_100_+_b_100),_range_query_with_1000_steps/Mimir-8                         831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum(a_100_+_b_100),_range_query_with_1000_steps/Prometheus-8                    831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum(a_2000_+_b_2000),_instant_query/Mimir-8                                     831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum(a_2000_+_b_2000),_instant_query/Prometheus-8                                831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum(a_2000_+_b_2000),_range_query_with_100_steps/Mimir-8                        831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum(a_2000_+_b_2000),_range_query_with_100_steps/Prometheus-8                   831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum(a_2000_+_b_2000),_range_query_with_1000_steps/Mimir-8                       831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum(a_2000_+_b_2000),_range_query_with_1000_steps/Prometheus-8                  831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(le)(rate(h_1[1m])),_instant_query/Mimir-8                               831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(le)(rate(h_1[1m])),_instant_query/Prometheus-8                          831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(le)(rate(h_1[1m])),_range_query_with_100_steps/Mimir-8                  831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(le)(rate(h_1[1m])),_range_query_with_100_steps/Prometheus-8             831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(le)(rate(h_1[1m])),_range_query_with_1000_steps/Mimir-8                 831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(le)(rate(h_1[1m])),_range_query_with_1000_steps/Prometheus-8            831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(le)(rate(h_100[1m])),_instant_query/Mimir-8                             831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(le)(rate(h_100[1m])),_instant_query/Prometheus-8                        831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(le)(rate(h_100[1m])),_range_query_with_100_steps/Mimir-8                831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(le)(rate(h_100[1m])),_range_query_with_100_steps/Prometheus-8           831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(le)(rate(h_100[1m])),_range_query_with_1000_steps/Mimir-8               831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(le)(rate(h_100[1m])),_range_query_with_1000_steps/Prometheus-8          831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(le)(rate(h_2000[1m])),_instant_query/Mimir-8                            831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(le)(rate(h_2000[1m])),_instant_query/Prometheus-8                       831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(le)(rate(h_2000[1m])),_range_query_with_100_steps/Mimir-8               831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(le)(rate(h_2000[1m])),_range_query_with_100_steps/Prometheus-8          831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(le)(rate(h_2000[1m])),_range_query_with_1000_steps/Mimir-8              831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
Query/sum_by_(le)(rate(h_2000[1m])),_range_query_with_1000_steps/Prometheus-8         831.9Ki ± 0%   818.0Ki ± 0%  -1.68% (p=0.002 n=6)
geomean                                                                               834.3Ki        820.3Ki       -1.67%
```
</details>

TL;RD The rate functions seem to have up to ~10% worse performance.

I suspect this is high enough to warrant maintaining two ring buffers. I really wanted to avoid doing so much duplication, especially because of the possibility of introducing a mistake, but unless you have another suggestion I will create a separate ring buffer.

Reply:
>I suspect this is high enough to warrant maintaining two ring buffers. I really wanted to avoid doing so much duplication, especially because of the possibility of introducing a mistake, but unless you have another suggestion I will create a separate ring buffer.

I think that's the best path here unfortunately. You should be able to rely on the existing tests to catch any regressions.